{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0351051e-741c-4204-898a-99632c70b433",
   "metadata": {
    "tags": []
   },
   "source": [
    "# variant annotation\n",
    "---\n",
    "\n",
    "**Author:** Chen Weng\n",
    "\n",
    "**Date:** Aug 13, 2024\n",
    "\n",
    "## Description\n",
    "Add more functions to annotate variants\n",
    "\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3bd1c82-358c-4381-baa1-96c8011e0a7d",
   "metadata": {},
   "source": [
    "# Load packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "37b6ba13-3a25-4684-b26f-8a516f894b34",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m Loading \u001b[34mredeemR\u001b[39m\n",
      "Warning message:\n",
      "“replacing previous import ‘ape::where’ by ‘dplyr::where’ when loading ‘redeemR’”\n",
      "Warning message:\n",
      "“replacing previous import ‘ape::keep.tip’ by ‘tidytree::keep.tip’ when loading ‘redeemR’”\n",
      "Warning message:\n",
      "“replacing previous import ‘ape::drop.tip’ by ‘tidytree::drop.tip’ when loading ‘redeemR’”\n",
      "Warning message:\n",
      "“replacing previous import ‘Biostrings::translate’ by ‘seqinr::translate’ when loading ‘dndscv’”\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n",
      "Found more than one class \"dist\" in cache; using the first, from namespace 'spam'\n",
      "\n",
      "Also defined by ‘BiocGenerics’\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## Load developing redeemR\n",
    "devtools::load_all(\"/lab/solexa_weissman/cweng/Packages/redeemR/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fe40ae3-f562-4350-bcf7-b743307fba8b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5c732b26-b91e-4398-98ae-2182484e60af",
   "metadata": {},
   "source": [
    "# Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ebb81144-4e76-4f38-be38-3749a8d6d9b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "data(haplogroup_markers)\n",
    "data(mitomap_freq)\n",
    "data(mito_diseases)\n",
    "data(mito_homopolymer)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f46e077-5896-4658-811e-cb1e5550d94a",
   "metadata": {},
   "source": [
    "# Load example redeemR object\n",
    "Read in any redeem object to test (recommend to load the variant after filter2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c670440f-0621-4991-8576-100ebb0688ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "Young1_HSC.redeemR_filter2 <- readRDS(\"/lab/solexa_weissman/cweng/Projects/Collaborator/Teng_Gao/MitoDrift/HSC/01_HSC_filter2_dir/Young1_HSC.redeemR_filter2.obj.rds\")$redeemR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69c1137e-2d35-4dfa-8168-ddd4162fd7e3",
   "metadata": {},
   "source": [
    "# Run wrap_up functions for variants annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d895f2ab-5102-4291-91d8-a83828cc6ad9",
   "metadata": {},
   "source": [
    "## run one line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c2fd87d9-2f75-458e-8b7a-cb6384768494",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Unknown columns: `ID`, `Locus`, `RSRS50`, `freq_all`, `freq_african`, `freq_asian`, `freq_eurasian`, `haplo_info`, `n_haplos`”\n",
      "[1] Loading the environment...\n",
      "\n",
      "[2] Annotating the mutations...\n",
      "\n",
      "Warning message in dndscv(mutations_all, refdb = mitorefdb, numcode = 2, max_muts_per_gene_per_sample = 5000, :\n",
      "“Mutations observed in contiguous sites within a sample. Please annotate or remove dinucleotide or complex substitutions for best results.”\n",
      "[3] Estimating global rates...\n",
      "\n",
      "[4] Running dNdSloc...\n",
      "\n",
      "[5] Running dNdScv...\n",
      "\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "“iteration limit reached”\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "“iteration limit reached”\n",
      "    Regression model for substitutions (theta = 3.83e+05).\n",
      "\n",
      "Warning message:\n",
      "“Unknown columns: `Disease`”\n"
     ]
    }
   ],
   "source": [
    "\n",
    "Young1_HSC.redeemR_filter2@V.fitered <- annotate_all_variants(Young1_HSC.redeemR_filter2@V.fitered)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4497965b-7d3b-40b5-aa9b-5657868c6b1c",
   "metadata": {},
   "source": [
    "## run separate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a7153d96-b926-4661-b9d0-577efc7a8749",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Unknown columns: `ID`, `haplo_info`”\n",
      "[1] Loading the environment...\n",
      "\n",
      "[2] Annotating the mutations...\n",
      "\n",
      "Warning message in dndscv(mutations_all, refdb = mitorefdb, numcode = 2, max_muts_per_gene_per_sample = 5000, :\n",
      "“Mutations observed in contiguous sites within a sample. Please annotate or remove dinucleotide or complex substitutions for best results.”\n",
      "Warning message in dndscv(mutations_all, refdb = mitorefdb, numcode = 2, max_muts_per_gene_per_sample = 5000, :\n",
      "“Same mutations observed in different sampleIDs. Please verify that these are independent events and remove duplicates otherwise.”\n",
      "[3] Estimating global rates...\n",
      "\n",
      "[4] Running dNdSloc...\n",
      "\n",
      "[5] Running dNdScv...\n",
      "\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "“iteration limit reached”\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "“iteration limit reached”\n",
      "    Regression model for substitutions (theta = 3.83e+05).\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Add hypermutable annotation\n",
    "Young1_HSC.redeemR_filter2@V.fitered <- annotate_variants_hypermutable(Young1_HSC.redeemR_filter2@V.fitered) \n",
    "\n",
    "# Add variant population statistics, gene, haplogroup markers¶\n",
    "Young1_HSC.redeemR_filter2@V.fitered <- annotate_variants_population_stats(Young1_HSC.redeemR_filter2@V.fitered)\n",
    "\n",
    "# Add black list positions\n",
    "Young1_HSC.redeemR_filter2@V.fitered <- annotate_variants_blacklist(Young1_HSC.redeemR_filter2@V.fitered)\n",
    "\n",
    "# Add homopolomer annotation\n",
    "Young1_HSC.redeemR_filter2@V.fitered <- annotate_variants_homopolymer(Young1_HSC.redeemR_filter2@V.fitered)\n",
    "\n",
    "# Add variant amino acid change and the gene\n",
    "Young1_HSC.redeemR_filter2@V.fitered <- annotate_variants_aachange(Young1_HSC.redeemR_filter2@V.fitered)\n",
    "\n",
    "# Add disease relevance information\n",
    "Young1_HSC.redeemR_filter2@V.fitered <- annotate_variants_mito_disease(Young1_HSC.redeemR_filter2@V.fitered)\n",
    "\n",
    "# Add nucleitide change\n",
    "Young1_HSC.redeemR_filter2@V.fitered <- Annotate_base_change(Young1_HSC.redeemR_filter2@V.fitered)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c76f0f15-8a79-4f68-9c1a-f194635f3e70",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5945deb8-d99c-471c-9c9c-e34b9da506b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 29</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>Variants</th><th scope=col>CellN</th><th scope=col>PositiveMean</th><th scope=col>PositiveMean_cts</th><th scope=col>maxcts</th><th scope=col>CellNPCT</th><th scope=col>TotalVcount</th><th scope=col>TotalCov</th><th scope=col>totalVAF</th><th scope=col>CV</th><th scope=col>⋯</th><th scope=col>freq_eurasian</th><th scope=col>n_haplos</th><th scope=col>blacklist_region</th><th scope=col>homopolymer</th><th scope=col>gene</th><th scope=col>aachange</th><th scope=col>impact</th><th scope=col>Disease</th><th scope=col>changes</th><th scope=col>types</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr[1d]&gt;</th><th scope=col>&lt;chr[1d]&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>6425_T_C</td><td>243</td><td>0.02250622</td><td> 1.111111</td><td>20</td><td>0.03899856</td><td> 274</td><td>351741</td><td>0.0007789823</td><td>0.04671960</td><td>⋯</td><td> 0.09</td><td>NA</td><td>                </td><td>NA</td><td>MT-CO1</td><td>P174P</td><td>Synonymous</td><td>NA                            </td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>146_T_C </td><td>127</td><td>0.34191850</td><td> 7.299213</td><td>31</td><td>0.02038196</td><td> 952</td><td>200949</td><td>0.0047375205</td><td>0.29552877</td><td>⋯</td><td>16.42</td><td>58</td><td>                </td><td>NA</td><td>NA    </td><td>NA   </td><td>NA        </td><td>Absence of Endometriosis      </td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>513_G_A </td><td> 89</td><td>0.02760451</td><td> 1.224719</td><td>13</td><td>0.01428342</td><td> 110</td><td>284935</td><td>0.0003860530</td><td>0.02784759</td><td>⋯</td><td> 1.18</td><td> 1</td><td>blacklist_region</td><td>NA</td><td>NA    </td><td>NA   </td><td>NA        </td><td>NA                            </td><td>G_A</td><td>transition</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>3244_G_A</td><td> 81</td><td>0.15582270</td><td> 8.481481</td><td>51</td><td>0.01299952</td><td> 689</td><td>443251</td><td>0.0015544240</td><td>0.27809690</td><td>⋯</td><td> 0.01</td><td>NA</td><td>                </td><td>NA</td><td>NA    </td><td>NA   </td><td>NA        </td><td>NA                            </td><td>G_A</td><td>transition</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>1348_G_A</td><td> 79</td><td>0.02079220</td><td> 1.126582</td><td> 6</td><td>0.01267854</td><td>  91</td><td>403824</td><td>0.0002253457</td><td>0.04595542</td><td>⋯</td><td>   NA</td><td>NA</td><td>                </td><td>NA</td><td>NA    </td><td>NA   </td><td>NA        </td><td>NA                            </td><td>G_A</td><td>transition</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>5913_G_A</td><td> 79</td><td>0.33335510</td><td>19.670886</td><td>88</td><td>0.01267854</td><td>1576</td><td>404987</td><td>0.0038914829</td><td>0.28282989</td><td>⋯</td><td> 1.29</td><td> 3</td><td>                </td><td>NA</td><td>MT-CO1</td><td>D4N  </td><td>Missense  </td><td>Prostate Cancer / hypertension</td><td>G_A</td><td>transition</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 29\n",
       "\\begin{tabular}{r|lllllllllllllllllllll}\n",
       "  & Variants & CellN & PositiveMean & PositiveMean\\_cts & maxcts & CellNPCT & TotalVcount & TotalCov & totalVAF & CV & ⋯ & freq\\_eurasian & n\\_haplos & blacklist\\_region & homopolymer & gene & aachange & impact & Disease & changes & types\\\\\n",
       "  & <chr> & <int> & <dbl> & <dbl> & <int> & <dbl> & <int> & <dbl> & <dbl> & <dbl> & ⋯ & <dbl> & <int> & <chr> & <chr> & <chr> & <chr{[}1d{]}> & <chr{[}1d{]}> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & 6425\\_T\\_C & 243 & 0.02250622 &  1.111111 & 20 & 0.03899856 &  274 & 351741 & 0.0007789823 & 0.04671960 & ⋯ &  0.09 & NA &                  & NA & MT-CO1 & P174P & Synonymous & NA                             & T\\_C & transition\\\\\n",
       "\t2 & 146\\_T\\_C  & 127 & 0.34191850 &  7.299213 & 31 & 0.02038196 &  952 & 200949 & 0.0047375205 & 0.29552877 & ⋯ & 16.42 & 58 &                  & NA & NA     & NA    & NA         & Absence of Endometriosis       & T\\_C & transition\\\\\n",
       "\t3 & 513\\_G\\_A  &  89 & 0.02760451 &  1.224719 & 13 & 0.01428342 &  110 & 284935 & 0.0003860530 & 0.02784759 & ⋯ &  1.18 &  1 & blacklist\\_region & NA & NA     & NA    & NA         & NA                             & G\\_A & transition\\\\\n",
       "\t4 & 3244\\_G\\_A &  81 & 0.15582270 &  8.481481 & 51 & 0.01299952 &  689 & 443251 & 0.0015544240 & 0.27809690 & ⋯ &  0.01 & NA &                  & NA & NA     & NA    & NA         & NA                             & G\\_A & transition\\\\\n",
       "\t5 & 1348\\_G\\_A &  79 & 0.02079220 &  1.126582 &  6 & 0.01267854 &   91 & 403824 & 0.0002253457 & 0.04595542 & ⋯ &    NA & NA &                  & NA & NA     & NA    & NA         & NA                             & G\\_A & transition\\\\\n",
       "\t6 & 5913\\_G\\_A &  79 & 0.33335510 & 19.670886 & 88 & 0.01267854 & 1576 & 404987 & 0.0038914829 & 0.28282989 & ⋯ &  1.29 &  3 &                  & NA & MT-CO1 & D4N   & Missense   & Prostate Cancer / hypertension & G\\_A & transition\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 29\n",
       "\n",
       "| <!--/--> | Variants &lt;chr&gt; | CellN &lt;int&gt; | PositiveMean &lt;dbl&gt; | PositiveMean_cts &lt;dbl&gt; | maxcts &lt;int&gt; | CellNPCT &lt;dbl&gt; | TotalVcount &lt;int&gt; | TotalCov &lt;dbl&gt; | totalVAF &lt;dbl&gt; | CV &lt;dbl&gt; | ⋯ ⋯ | freq_eurasian &lt;dbl&gt; | n_haplos &lt;int&gt; | blacklist_region &lt;chr&gt; | homopolymer &lt;chr&gt; | gene &lt;chr&gt; | aachange &lt;chr[1d]&gt; | impact &lt;chr[1d]&gt; | Disease &lt;chr&gt; | changes &lt;chr&gt; | types &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | 6425_T_C | 243 | 0.02250622 |  1.111111 | 20 | 0.03899856 |  274 | 351741 | 0.0007789823 | 0.04671960 | ⋯ |  0.09 | NA | <!----> | NA | MT-CO1 | P174P | Synonymous | NA                             | T_C | transition |\n",
       "| 2 | 146_T_C  | 127 | 0.34191850 |  7.299213 | 31 | 0.02038196 |  952 | 200949 | 0.0047375205 | 0.29552877 | ⋯ | 16.42 | 58 | <!----> | NA | NA     | NA    | NA         | Absence of Endometriosis       | T_C | transition |\n",
       "| 3 | 513_G_A  |  89 | 0.02760451 |  1.224719 | 13 | 0.01428342 |  110 | 284935 | 0.0003860530 | 0.02784759 | ⋯ |  1.18 |  1 | blacklist_region | NA | NA     | NA    | NA         | NA                             | G_A | transition |\n",
       "| 4 | 3244_G_A |  81 | 0.15582270 |  8.481481 | 51 | 0.01299952 |  689 | 443251 | 0.0015544240 | 0.27809690 | ⋯ |  0.01 | NA | <!----> | NA | NA     | NA    | NA         | NA                             | G_A | transition |\n",
       "| 5 | 1348_G_A |  79 | 0.02079220 |  1.126582 |  6 | 0.01267854 |   91 | 403824 | 0.0002253457 | 0.04595542 | ⋯ |    NA | NA | <!----> | NA | NA     | NA    | NA         | NA                             | G_A | transition |\n",
       "| 6 | 5913_G_A |  79 | 0.33335510 | 19.670886 | 88 | 0.01267854 | 1576 | 404987 | 0.0038914829 | 0.28282989 | ⋯ |  1.29 |  3 | <!----> | NA | MT-CO1 | D4N   | Missense   | Prostate Cancer / hypertension | G_A | transition |\n",
       "\n"
      ],
      "text/plain": [
       "  Variants CellN PositiveMean PositiveMean_cts maxcts CellNPCT   TotalVcount\n",
       "1 6425_T_C 243   0.02250622    1.111111        20     0.03899856  274       \n",
       "2 146_T_C  127   0.34191850    7.299213        31     0.02038196  952       \n",
       "3 513_G_A   89   0.02760451    1.224719        13     0.01428342  110       \n",
       "4 3244_G_A  81   0.15582270    8.481481        51     0.01299952  689       \n",
       "5 1348_G_A  79   0.02079220    1.126582         6     0.01267854   91       \n",
       "6 5913_G_A  79   0.33335510   19.670886        88     0.01267854 1576       \n",
       "  TotalCov totalVAF     CV         ⋯ freq_eurasian n_haplos blacklist_region\n",
       "1 351741   0.0007789823 0.04671960 ⋯  0.09         NA                       \n",
       "2 200949   0.0047375205 0.29552877 ⋯ 16.42         58                       \n",
       "3 284935   0.0003860530 0.02784759 ⋯  1.18          1       blacklist_region\n",
       "4 443251   0.0015544240 0.27809690 ⋯  0.01         NA                       \n",
       "5 403824   0.0002253457 0.04595542 ⋯    NA         NA                       \n",
       "6 404987   0.0038914829 0.28282989 ⋯  1.29          3                       \n",
       "  homopolymer gene   aachange impact     Disease                        changes\n",
       "1 NA          MT-CO1 P174P    Synonymous NA                             T_C    \n",
       "2 NA          NA     NA       NA         Absence of Endometriosis       T_C    \n",
       "3 NA          NA     NA       NA         NA                             G_A    \n",
       "4 NA          NA     NA       NA         NA                             G_A    \n",
       "5 NA          NA     NA       NA         NA                             G_A    \n",
       "6 NA          MT-CO1 D4N      Missense   Prostate Cancer / hypertension G_A    \n",
       "  types     \n",
       "1 transition\n",
       "2 transition\n",
       "3 transition\n",
       "4 transition\n",
       "5 transition\n",
       "6 transition"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(Young1_HSC.redeemR_filter2@V.fitered)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "291539ef-2290-4a8b-bea3-b1c02838cb39",
   "metadata": {},
   "source": [
    "# Breakdown each part "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bc268d3-3693-40ba-b8a0-31e814fcf2de",
   "metadata": {},
   "source": [
    "## Add variant population statistics, gene, haplogroup markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "27885318-5b70-48f6-a4b8-76173e092d6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_annotation <- Young1_HSC.redeemR_filter2@V.fitered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "26e012f0-bd6b-4596-87ad-04e278c03141",
   "metadata": {},
   "outputs": [],
   "source": [
    "data(mitomap_freq)\n",
    "mitomap_freq <-mitomap_freq[,c(\"ID\",\"Locus\", \"RSRS50\",\"Overall.Variant.Frequency.in.Sequence.Set\", \"Frequency.in.Lineage.L\", \"Frequency.in.Lineage.M\", \"Frequency.in.Lineage.N\")]\n",
    "names(mitomap_freq)[4:7] <-c(\"freq_all\",\"freq_african\",\"freq_asian\",\"freq_eurasian\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "f2383e0a-85d1-4cba-89c3-2163476f4a8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "data(haplogroup_markers)\n",
    "haplogroup_markers_summary <- haplogroup_markers %>%\n",
    "  group_by(ID) %>%\n",
    "  summarize(\n",
    "    haplo_info = paste0(\n",
    "      paste(haplo, collapse = \"_\"),  # collapse haplo by “_”\n",
    "      \";\",\n",
    "      paste(freq, collapse = \"_\")    # collapse freq by “_”\n",
    "    ),\n",
    "    .groups = \"drop\",\n",
    "    n_haplos = n()\n",
    "  ) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "cb621a89-1749-4181-9079-468876883225",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Unknown columns: `ID`, `Locus`, `RSRS50`, `freq_all`, `freq_african`, `freq_asian`, `freq_eurasian`, `haplo_info`, `n_haplos`”\n"
     ]
    }
   ],
   "source": [
    "variant_annotation <- variant_annotation %>% \n",
    "                    select(-one_of(unique(c(names(mitomap_freq),names(haplogroup_markers_summary))))) %>%\n",
    "                    left_join(mitomap_freq, by = c(\"Variants\" = \"ID\")) %>%\n",
    "                    arrange(desc(CellN)) %>% \n",
    "                    left_join(haplogroup_markers_summary[, c(\"ID\",\"n_haplos\")], by = c(\"Variants\" = \"ID\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566c56fe-39c4-4c8a-b160-c894f5f56a61",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a590568a-26ea-4de2-add4-3dd3de98655f",
   "metadata": {},
   "source": [
    "> Wrap these up annotate_variants_population_stats and update in redeemR Package variant_annotation.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "696adad4-67de-490d-bec7-3b67008bd268",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Annotate variant table with population frequency and haplogroup statistics\n",
    "#'\n",
    "#' @description\n",
    "#' Adds mitochondrial variant information from \\code{mitomap_freq} and \n",
    "#' haplogroup marker counts from \\code{haplogroup_markers} to a variant annotation\n",
    "#' data frame.  data. It drops any existing columns that would collide, then merges in:\n",
    "#' \\itemize{\n",
    "#'   \\item{\\code{Locus}}{Gene name, tRNA, or other feature from \\code{mitomap_freq}.}\n",
    "#'   \\item{\\code{RSRS50}}{Logical flag indicating whether the variant is an ancestral (RSRS50) variant.}\n",
    "#'   \\item{\\code{freq_all, freq_african, freq_asian, freq_eurasian}}{Mitochondrial variant frequencies in the overall, African, Asian, and Eurasian populations, respectively.}\n",
    "#'   \\item{\\code{n_haplos}}{Number of population‐level ancestry branches (haplogroups) that consider this variant a marker (posterior >80\\%).}\n",
    "#' }\n",
    "#'\n",
    "#' @param variant_annotation A data.frame or tibble containing at least columns \\code{Variants} and \\code{CellN}.\n",
    "#'\n",
    "#' @return The input \\code{variant_annotation}, augmented with:\n",
    "#' \\describe{\n",
    "#'   \\item{\\code{Locus}}{Feature name (gene/tRNA/etc).}\n",
    "#'   \\item{\\code{RSRS50}}{Ancestral variant flag (logical).}\n",
    "#'   \\item{\\code{freq_all}}{Overall variant frequency.}\n",
    "#'   \\item{\\code{freq_african}}{Variant frequency in African population.}\n",
    "#'   \\item{\\code{freq_asian}}{Variant frequency in Asian population.}\n",
    "#'   \\item{\\code{freq_eurasian}}{Variant frequency in Eurasian population.}\n",
    "#'   \\item{\\code{n_haplos}}{Count of haplogroups (ancestry branches) marking this variant (>80\\% posterior).}\n",
    "#' }\n",
    "#'\n",
    "#' @importFrom dplyr select one_of left_join arrange desc group_by summarize\n",
    "#' @export\n",
    "annotate_variants_population_stats <- function(variant_annotation){ \n",
    "    data(mitomap_freq)\n",
    "    data(haplogroup_markers)\n",
    "    mitomap_freq <-mitomap_freq[,c(\"ID\",\"Locus\", \"RSRS50\",\"Overall.Variant.Frequency.in.Sequence.Set\", \"Frequency.in.Lineage.L\", \"Frequency.in.Lineage.M\", \"Frequency.in.Lineage.N\")]\n",
    "    names(mitomap_freq)[4:7] <-c(\"freq_all\",\"freq_african\",\"freq_asian\",\"freq_eurasian\")\n",
    "    haplogroup_markers_summary <- haplogroup_markers %>%\n",
    "      group_by(ID) %>%\n",
    "      summarize(\n",
    "        haplo_info = paste0(\n",
    "          paste(haplo, collapse = \"_\"),  # collapse haplo by “_”\n",
    "          \";\",\n",
    "          paste(freq, collapse = \"_\")    # collapse freq by “_”\n",
    "        ),\n",
    "        .groups = \"drop\",\n",
    "        n_haplos = n()\n",
    "      ) \n",
    "    variant_annotation <- variant_annotation %>% \n",
    "                        select(-one_of(unique(c(names(mitomap_freq),names(haplogroup_markers_summary))))) %>%\n",
    "                        left_join(mitomap_freq, by = c(\"Variants\" = \"ID\")) %>%\n",
    "                        arrange(desc(CellN)) %>% \n",
    "                        left_join(haplogroup_markers_summary[, c(\"ID\",\"n_haplos\")], by = c(\"Variants\" = \"ID\"))\n",
    "\n",
    "    return(variant_annotation)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "743da8b7-71c2-4b8c-8444-e6c00250947a",
   "metadata": {},
   "source": [
    "## Add black list positions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a34cb40b-1690-4c3b-8ab3-d8baeb470227",
   "metadata": {},
   "source": [
    "(1)\n",
    "Misalignment due to ACCCCCCCTCCCCC (rCRS 302-315)\n",
    "\n",
    "(2)\n",
    "Misalignment due to GCACACACACACC (rCRS 513-525)\n",
    "\n",
    "(3)\n",
    "Misalignment due to 3107N in rCRS (rCRS 3,105-3,109)\n",
    "\n",
    "(4)\n",
    "Misalignment due to ACCCCC (rCRS 16,182-16,187)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "bc526821-57ad-40a3-a0a9-d13c4b9ebcbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "black_list_regions<-c(302:315, 513:525, 3105:3109, 16182:16187)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "d1593857-1602-4017-bb77-105429d8d00d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pos<- as.numeric(strsplit(variant_annotation$Variants, \"_\") %>% sapply(function(x){x[1]}))\n",
    "variant_annotation$blacklist_region<- ifelse(pos %in% black_list_regions,\"blacklist_region\",\"\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a71f838e-09ab-43cd-a122-320737769c2e",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cdaee26-0b7f-4b46-8ed6-4817728f8d73",
   "metadata": {},
   "source": [
    "> wrap these up and update in redeemR Package as annotate_variants_blacklist "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "1bcc392d-0e08-47c6-a5b9-2a30a598cb17",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Annotate variants with mitochondrial blacklist region flag\n",
    "#'\n",
    "#' @description\n",
    "#' Flags variants whose genomic position falls within known problematic\n",
    "#' regions of the mitochondrial genome (blacklist regions).\n",
    "#'\n",
    "#' @details\n",
    "#' The blacklist intervals correspond to regions in the revised Cambridge Reference\n",
    "#' Sequence (rCRS) prone to misalignment:\n",
    "#' \\enumerate{\n",
    "#'   \\item Misalignment due to \\code{ACCCCCC T C C C C C C} (rCRS 302–315)\n",
    "#'   \\item Misalignment due to \\code{G C A C A C A C A C A C C} (rCRS 513–525)\n",
    "#'   \\item Misalignment due to \\code{3107N} ambiguity in rCRS (rCRS 3105–3109)\n",
    "#'   \\item Misalignment due to \\code{ACCCCC} homopolymer (rCRS 16182–16187)\n",
    "#' }\n",
    "#'\n",
    "#' @param variant_annotation A data.frame or tibble containing at least a column\n",
    "#'   named \\code{Variants}, where each entry is of the form \\code{\"<pos>_<ref>_<alt>\"}.\n",
    "#'\n",
    "#' @return The input \\code{variant_annotation} with an added column:\n",
    "#' \\describe{\n",
    "#'   \\item{\\code{blacklist_region}}{Character; \\code{\"blacklist_region\"} if the variant\n",
    "#'     position is in one of the predefined blacklist intervals, otherwise \\code{\"\"}.}\n",
    "#' }\n",
    "#'\n",
    "#' @export\n",
    "annotate_variants_blacklist <- function(variant_annotation){\n",
    "    black_list_regions<-c(302:315, 513:525, 3105:3109, 16182:16187)\n",
    "    pos<- as.numeric(strsplit(variant_annotation$Variants, \"_\") %>% sapply(function(x){x[1]}))\n",
    "    variant_annotation$blacklist_region<- ifelse(pos %in% black_list_regions,\"blacklist_region\",\"\")\n",
    "    return(variant_annotation)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e113839-3a8e-4b10-b7e6-e926766002d8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "629c8411-4eab-4b19-a055-3219663fa677",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Add homopolomer annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "id": "a8088337-ba52-4caf-8c77-16417bc31f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each pos, find the first interval that contains it\n",
    "pos<- as.numeric(strsplit(variant_annotation$Variants, \"_\") %>% sapply(function(x){x[1]}))\n",
    "assigned <- sapply(pos, function(pos) {\n",
    "  idx <- which(mito_homopolymer$V2 <= pos & pos <= mito_homopolymer$V3)\n",
    "  if (length(idx)) mito_homopolymer$V4[idx[1]] else NA\n",
    "})\n",
    "\n",
    "# assemble result\n",
    "pos_homopolymer <- data.frame(\n",
    "  position = pos,\n",
    "  V4        = assigned,\n",
    "  stringsAsFactors = FALSE\n",
    ")\n",
    "\n",
    "variant_annotation <- variant_annotation %>% mutate(homopolymer = pos_homopolymer$V4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14286b4b-46af-4055-9a3d-bd62fa33fcc9",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "447b7b2c-5409-4661-a36d-5a2d96e3a8ba",
   "metadata": {
    "tags": []
   },
   "source": [
    "> wrap up above and update in redeemR Package as annotate_variants_homopolymer "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "id": "bef7d6b0-5bd8-46a2-9146-041d13ba44d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Annotate variants with mitochondrial homopolymer context\n",
    "#'\n",
    "#' @description\n",
    "#' Flags each variant by the homopolymer region in which its position falls,\n",
    "#' based on predefined mitochondrial homopolymer intervals.\n",
    "#'\n",
    "#' @details\n",
    "#' The function loads the \\code{mito_homopolymer} data frame, which must contain\n",
    "#' columns \\code{V2} (start), \\code{V3} (end), and \\code{V4} (region label).\n",
    "#' For each variant position, it finds the first homopolymer interval that\n",
    "#' contains that position and assigns its label; otherwise, \\code{NA}.\n",
    "#'\n",
    "#' @param variant_annotation A data.frame or tibble containing at least a column\n",
    "#'   named \\code{Variants}, where each entry is of the form \\code{\"<pos>_<ref>_<alt>\"}.\n",
    "#'\n",
    "#' @return The input \\code{variant_annotation} with an added column:\n",
    "#' \\describe{\n",
    "#'   \\item{\\code{homopolymer}}{Character; the label of the homopolymer region\n",
    "#'     (from \\code{mito_homopolymer\\$V4}) containing the variant position, or\n",
    "#'     \\code{NA} if none.}\n",
    "#' }\n",
    "#'\n",
    "#' @importFrom dplyr mutate\n",
    "#' @export\n",
    "annotate_variants_homopolymer <- function(variant_annotation){\n",
    "    data(mito_homopolymer)   \n",
    "    pos<- as.numeric(strsplit(variant_annotation$Variants, \"_\") %>% sapply(function(x){x[1]}))\n",
    "    # for each pos, find the first interval that contains it\n",
    "    assigned <- sapply(pos, function(pos) {\n",
    "      idx <- which(mito_homopolymer$V2 <= pos & pos <= mito_homopolymer$V3)\n",
    "      if (length(idx)) mito_homopolymer$V4[idx[1]] else NA\n",
    "    })\n",
    "\n",
    "    # assemble result\n",
    "    pos_homopolymer <- data.frame(\n",
    "      position = pos,\n",
    "      V4        = assigned,\n",
    "      stringsAsFactors = FALSE\n",
    "    )\n",
    "    variant_annotation <- variant_annotation %>% mutate(homopolymer = pos_homopolymer$V4)\n",
    "    return(variant_annotation)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b20c4462-8005-45f5-9a7d-c3ed8e2a3863",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "88832623-3d6f-4289-9443-eee45a01ccc8",
   "metadata": {},
   "source": [
    "## Add variant amino acid change and the gene\n",
    "(optional) <br/>\n",
    "Dependency: library(https://github.com/im3sanger/dndscv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "a75b143c-2a25-4a44-afbe-207edf7a9188",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Parse mitochondrial variants into a tidy table\n",
    "#'\n",
    "#' @description\n",
    "#' Internal helper to split variant strings of the form `\"<pos>_<ref>_<mut>\"`\n",
    "#' into a data frame with columns for sample ID, chromosome, position, reference,\n",
    "#' and alternate allele.\n",
    "#'\n",
    "#' @param variants Character vector of variant strings, e.g. `\"123_A_G\"`.\n",
    "#' @param ID       Character scalar; sample identifier to assign to all rows.\n",
    "#'\n",
    "#' @return A data.frame with columns:\n",
    "#'\n",
    "#' @keywords internal\n",
    "#' @noRd\n",
    "make_mutation_table<-function(variants,ID=\"\"){\n",
    "    mito_mutations<-strsplit(variants,\"_\") %>% do.call(rbind,.) %>% as.data.frame()\n",
    "    mito_mutations<-data.frame(sampleID=ID,chr=\"MT\",mito_mutations)\n",
    "    names(mito_mutations)[3:5]<-c(\"pos\",\"ref\",\"mut\")\n",
    "    mito_mutations$pos<-as.integer(mito_mutations$pos)\n",
    "    return(mito_mutations)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "b47fae64-0e8d-4b32-8fed-b25d0aeab446",
   "metadata": {},
   "outputs": [],
   "source": [
    "require(dndscv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "93ff3cdd-b672-4107-80e7-1b53441f7697",
   "metadata": {},
   "outputs": [],
   "source": [
    "mitorefdb <- system.file(\"data\", \"GRCH38_MT_bioMRT_output_refcds.rda\",package = \"redeemR\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "id": "1d329c3d-ed70-4e7b-8538-89993bb8e060",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_all<-make_mutation_table(variant_annotation$Variants,\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "cb747358-c4d4-44ef-bdbd-578f724c9a26",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[1] Loading the environment...\n",
      "\n",
      "[2] Annotating the mutations...\n",
      "\n",
      "Warning message in dndscv(mutations_all, refdb = mitorefdb, numcode = 2, max_muts_per_gene_per_sample = 5000, :\n",
      "“Mutations observed in contiguous sites within a sample. Please annotate or remove dinucleotide or complex substitutions for best results.”\n",
      "[3] Estimating global rates...\n",
      "\n",
      "[4] Running dNdSloc...\n",
      "\n",
      "[5] Running dNdScv...\n",
      "\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "“iteration limit reached”\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "“iteration limit reached”\n",
      "    Regression model for substitutions (theta = 3.83e+05).\n",
      "\n"
     ]
    }
   ],
   "source": [
    "mito_dndsout_all = dndscv(mutations_all,refdb=mitorefdb,numcode = 2,max_muts_per_gene_per_sample = 5000,max_coding_muts_per_sample = 5000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "d3af2e36-2c10-4098-a6e2-769555bfcaa7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "mito_aa_annotation<- mito_dndsout_all$annotmuts %>% mutate(Variants= paste(pos, ref, mut, sep=\"_\")) %>%\n",
    "select(Variants, gene, aachange, impact)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "b3c0e9f0-0828-4953-8296-ffc194915160",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_annotation <- variant_annotation %>% select(-any_of(c(\"aachange\", \"impact\"))) %>% left_join(mito_aa_annotation,by = \"Variants\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "731ddb22-30dd-42f6-bbb1-eb5d00d10910",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6a4d837-d3a5-431b-aedb-c53316285a67",
   "metadata": {},
   "source": [
    "> wrap up above and update in redeemR Package as annotate_variants_aachange "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "id": "67cfa71c-405d-487e-a8cc-e86a9655aeb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Annotate variants with amino-acid change and predicted impact via dndscv\n",
    "#'\n",
    "#' @description\n",
    "#' Runs dndscv on the mitochondrial variants to obtain gene-level\n",
    "#' annotation of amino-acid changes and their predicted impact.\n",
    "#'\n",
    "#' @details\n",
    "#' This internal function:\n",
    "#' 1. Loads the precomputed mitochondrial CDS reference database from the\n",
    "#'    redeemR package (`GRCH38_MT_bioMRT_output_refcds.rda`).\n",
    "#' 2. Converts `Variants` (\"<pos>_<ref>_<alt>\") into the 5-column table\n",
    "#'    required by dndscv using `make_mutation_table`.\n",
    "#' 3. Runs `dndscv()` with permissive mutation limits to annotate all coding\n",
    "#'    changes.\n",
    "#' 4. Extracts and merges back the `gene`, `aachange`, and `impact` fields\n",
    "#'    for each variant.\n",
    "#'\n",
    "#' @param variant_annotation A data.frame or tibble containing at least a column\n",
    "#'   named `Variants` of the form \"<pos>_<ref>_<alt>\".\n",
    "#'\n",
    "#' @return The input `variant_annotation`, augmented with columns:\n",
    "#' \\describe{\n",
    "#'   \\item{gene}{Gene symbol where the variant falls.}\n",
    "#'   \\item{aachange}{Amino-acid change notation (e.g. p.W88C).}\n",
    "#'   \\item{impact}{Predicted functional impact category from dndscv.}\n",
    "#' }\n",
    "#'\n",
    "#' @importFrom dndscv dndscv\n",
    "#' @importFrom dplyr mutate select left_join any_of\n",
    "#' @export\n",
    "annotate_variants_aachange <- function(variant_annotation){\n",
    "    require(dndscv)\n",
    "    mitorefdb <- system.file(\"data\", \"GRCH38_MT_bioMRT_output_refcds.rda\",package = \"redeemR\")\n",
    "    mutations_all<- make_mutation_table(variant_annotation$Variants,\"\")\n",
    "    mito_dndsout_all <- dndscv(mutations_all,refdb=mitorefdb,numcode = 2,max_muts_per_gene_per_sample = 5000,max_coding_muts_per_sample = 5000)\n",
    "    mito_aa_annotation<- mito_dndsout_all$annotmuts %>% \n",
    "                         mutate(Variants= paste(pos, ref, mut, sep=\"_\")) %>%\n",
    "                         select(Variants, gene, aachange, impact)\n",
    "    variant_annotation <- variant_annotation %>% select(-any_of(c(\"aachange\", \"impact\"))) %>% left_join(mito_aa_annotation,by = \"Variants\")\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fab1aeac-04d8-42a2-89e0-b6c9e7e57f12",
   "metadata": {},
   "source": [
    "## Add disease relevance information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fd51d974-af8c-4218-9b72-ddcf67f7b25f",
   "metadata": {},
   "outputs": [],
   "source": [
    "data(mito_diseases)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b2eda06e-f4f2-411b-b6d3-d49be3aa63a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Unknown columns: `Disease`”\n"
     ]
    }
   ],
   "source": [
    "variant_annotation <- variant_annotation %>% \n",
    "                    select(-one_of(unique(c(\"Disease\")))) %>%\n",
    "                    left_join(mito_diseases[,c(\"ID\",\"Disease\")], by = c(\"Variants\" = \"ID\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94a957b0-44a8-43c4-ba8b-e3c614da3df9",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8484ac7b-fe27-44ca-b271-9ce9660db7ac",
   "metadata": {},
   "source": [
    "> wrap up for function annotate_variants_mito_disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "95ec2e4f-9285-4077-8438-cf2da046c6d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Annotate variants with mitochondrial disease associations\n",
    "#'\n",
    "#' @description\n",
    "#' Adds disease annotations from the \\code{mito_diseases} dataset for each variant.\n",
    "#'\n",
    "#' @param variant_annotation A data.frame or tibble containing at least a column\n",
    "#'   named \\code{Variants}.\n",
    "#'\n",
    "#' @return The input \\code{variant_annotation} with an added column:\n",
    "#' \\describe{\n",
    "#'   \\item{Disease}{Disease association from \\code{mito_diseases}, or \\code{NA} if none.}\n",
    "#' }\n",
    "#'\n",
    "#' @importFrom dplyr select one_of left_join\n",
    "#' @export\n",
    "#' Annotate variants with mitochondrial disease associations\n",
    "#'\n",
    "#' @description\n",
    "#' Adds disease annotations from the mito_diseases dataset for each variant.\n",
    "#'\n",
    "#' @param variant_annotation A data.frame or tibble containing at least a column named \\code{Variants}.\n",
    "#'\n",
    "#' @return The input \\code{variant_annotation} with an added column:\n",
    "#' \\describe{\\item{Disease}{Disease association(s) from \\code{mito_diseases}, NA if none.}}\n",
    "#' @importFrom dplyr select one_of left_join\n",
    "#' @export\n",
    "annotate_variants_mito_disease <- function(variant_annotation){\n",
    "    data(mito_diseases)\n",
    "    variant_annotation <- variant_annotation %>% \n",
    "                    select(-one_of(unique(c(\"Disease\")))) %>%\n",
    "                    left_join(mito_diseases[,c(\"ID\",\"Disease\")], by = c(\"Variants\" = \"ID\"))\n",
    "    return(variant_annotation)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffc27c5a-b114-4a0c-b352-30c2b842c728",
   "metadata": {},
   "source": [
    "# Update the variant summary statistics from mtx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e45fd7f-6907-46fb-b9e3-f001bd4d6cd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#' Create variant summary table from count matrix\n",
    "#'\n",
    "#' @description\n",
    "#' Summarizes a variant-by-cell count matrix to produce per-variant metrics.\n",
    "#'\n",
    "#' @details\n",
    "#' Calculates for each variant (column of \\code{mtx}):\n",
    "#' \\itemize{\n",
    "#'   \\item{\\code{cellN}}{Number of cells with count > 0.}\n",
    "#'   \\item{\\code{max_count}}{Maximum count observed across cells.}\n",
    "#'   \\item{\\code{mean_count}}{Mean of non-zero counts.}\n",
    "#'   \\item{\\code{ct1_cellN}}{Number of cells with count == 1.}\n",
    "#'   \\item{\\code{ct2_cellN}}{Number of cells with count == 2.}\n",
    "#'   \\item{\\code{ct3plus_cellN}}{Number of cells with count > 2.}\n",
    "#'   \\item{\\code{pct_ct2plus}}{Fraction of cells with count >= 2.}\n",
    "#'   \\item{\\code{cellNPCT}}{Fraction of cells with count > 0 relative to total cells.}\n",
    "#' }\n",
    "#'\n",
    "#' @param mtx Numeric matrix with variants as columns and cells as rows.\n",
    "#'\n",
    "#' @return A data.frame with one row per variant, including the metrics above and a \n",
    "#'         column \\code{Variants} with variant identifiers (column names of \\code{mtx}).\n",
    "#'\n",
    "#' @importFrom tibble rownames_to_column\n",
    "#' @importFrom dplyr mutate\n",
    "#' @export\n",
    "make_V.fitered_from_mtx<- function(mtx){\n",
    "    mtx_summarystat<-data.frame(CellN=apply(mtx, 2, function(x){sum(x>0)}),\n",
    "    max_count=apply(mtx, 2, max),\n",
    "    mean_count=apply(mtx, 2, function(x){mean(x[x>0])}),\n",
    "    ct1_cellN=apply(mtx, 2, function(x){length(x[x==1])}),\n",
    "    ct2_cellN=apply(mtx, 2, function(x){length(x[x==2])}),\n",
    "    ct3plus_cellN=apply(mtx, 2, function(x){length(x[x>2])})                  \n",
    "    ) %>% \n",
    "    tibble::rownames_to_column(\"Variants\") %>%\n",
    "    mutate(Variants = convert_variant(Variants), pct_ct2plus=(ct2_cellN+ct3plus_cellN)/CellN) %>% mutate(cellNPCT=CellN/nrow(mtx))\n",
    "    return(mtx_summarystat)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bcb13117-6861-4465-8b36-06ce8434bbc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_summary_stat_from_mtx <- make_V.fitered_from_mtx(Young1_HSC.redeemR_filter2@Cts.Mtx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "75e29f46-f3af-4a50-b83a-387ef549068b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Unknown columns: `ID`, `Locus`, `RSRS50`, `freq_all`, `freq_african`, `freq_asian`, `freq_eurasian`, `haplo_info`, `n_haplos`”\n",
      "[1] Loading the environment...\n",
      "\n",
      "[2] Annotating the mutations...\n",
      "\n",
      "Warning message in dndscv(mutations_all, refdb = mitorefdb, numcode = 2, max_muts_per_gene_per_sample = 5000, :\n",
      "“Mutations observed in contiguous sites within a sample. Please annotate or remove dinucleotide or complex substitutions for best results.”\n",
      "[3] Estimating global rates...\n",
      "\n",
      "[4] Running dNdSloc...\n",
      "\n",
      "[5] Running dNdScv...\n",
      "\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "“iteration limit reached”\n",
      "Warning message in theta.ml(Y, mu, sum(w), w, limit = control$maxit, trace = control$trace > :\n",
      "“iteration limit reached”\n",
      "    Regression model for substitutions (theta = 3.83e+05).\n",
      "\n",
      "Warning message:\n",
      "“Unknown columns: `Disease`”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3782 × 25</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Variants</th><th scope=col>CellN</th><th scope=col>max_count</th><th scope=col>mean_count</th><th scope=col>ct1_cellN</th><th scope=col>ct2_cellN</th><th scope=col>ct3plus_cellN</th><th scope=col>pct_ct2plus</th><th scope=col>cellNPCT</th><th scope=col>hyper_label</th><th scope=col>⋯</th><th scope=col>freq_eurasian</th><th scope=col>n_haplos</th><th scope=col>blacklist_region</th><th scope=col>homopolymer</th><th scope=col>gene</th><th scope=col>aachange</th><th scope=col>impact</th><th scope=col>Disease</th><th scope=col>changes</th><th scope=col>types</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>⋯</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr[1d]&gt;</th><th scope=col>&lt;chr[1d]&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>1000_T_C </td><td> 3</td><td> 4</td><td> 2.000000</td><td>2</td><td>0</td><td>1</td><td>0.3333333</td><td>0.0005104645</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10005_A_G</td><td> 2</td><td> 3</td><td> 2.000000</td><td>1</td><td>0</td><td>1</td><td>0.5000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>0.02</td><td>NA</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>A_G</td><td>transition</td></tr>\n",
       "\t<tr><td>10007_T_C</td><td> 2</td><td> 3</td><td> 2.500000</td><td>0</td><td>1</td><td>1</td><td>1.0000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>0.10</td><td>NA</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10009_G_A</td><td> 2</td><td>10</td><td> 5.500000</td><td>1</td><td>0</td><td>1</td><td>0.5000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>G_A</td><td>transition</td></tr>\n",
       "\t<tr><td>10010_T_C</td><td> 3</td><td>12</td><td> 4.666667</td><td>2</td><td>0</td><td>1</td><td>0.3333333</td><td>0.0005104645</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10020_T_C</td><td> 6</td><td>13</td><td> 4.833333</td><td>1</td><td>2</td><td>3</td><td>0.8333333</td><td>0.0010209290</td><td></td><td>⋯</td><td>0.01</td><td>NA</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10031_T_C</td><td> 3</td><td> 2</td><td> 1.333333</td><td>2</td><td>1</td><td>0</td><td>0.3333333</td><td>0.0005104645</td><td></td><td>⋯</td><td>0.18</td><td> 2</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>1004_G_A </td><td> 4</td><td> 3</td><td> 2.000000</td><td>2</td><td>0</td><td>2</td><td>0.5000000</td><td>0.0006806194</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>G_A</td><td>transition</td></tr>\n",
       "\t<tr><td>10045_T_C</td><td> 2</td><td>32</td><td>16.500000</td><td>1</td><td>0</td><td>1</td><td>0.5000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>0.04</td><td>NA</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>1005_T_C </td><td> 6</td><td> 7</td><td> 2.000000</td><td>5</td><td>0</td><td>1</td><td>0.1666667</td><td>0.0010209290</td><td></td><td>⋯</td><td>0.62</td><td> 8</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10054_G_A</td><td> 7</td><td>11</td><td> 2.428571</td><td>6</td><td>0</td><td>1</td><td>0.1428571</td><td>0.0011910839</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>G_A</td><td>transition</td></tr>\n",
       "\t<tr><td>1006_T_C </td><td> 9</td><td>30</td><td> 4.555556</td><td>7</td><td>0</td><td>2</td><td>0.2222222</td><td>0.0015313936</td><td></td><td>⋯</td><td>0.00</td><td>NA</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10063_A_G</td><td> 4</td><td> 4</td><td> 2.750000</td><td>1</td><td>1</td><td>2</td><td>0.7500000</td><td>0.0006806194</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-ND3</td><td>N2S </td><td>Missense  </td><td>NA</td><td>A_G</td><td>transition</td></tr>\n",
       "\t<tr><td>10066_T_C</td><td> 4</td><td> 5</td><td> 2.000000</td><td>3</td><td>0</td><td>1</td><td>0.2500000</td><td>0.0006806194</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-ND3</td><td>F3S </td><td>Missense  </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10068_G_A</td><td>10</td><td> 4</td><td> 1.900000</td><td>6</td><td>1</td><td>3</td><td>0.4000000</td><td>0.0017015484</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-ND3</td><td>A4T </td><td>Missense  </td><td>NA</td><td>G_A</td><td>transition</td></tr>\n",
       "\t<tr><td>10071_T_C</td><td> 4</td><td> 2</td><td> 1.250000</td><td>3</td><td>1</td><td>0</td><td>0.2500000</td><td>0.0006806194</td><td></td><td>⋯</td><td>0.00</td><td>NA</td><td></td><td>NA     </td><td>MT-ND3</td><td>L5L </td><td>Synonymous</td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10075_T_C</td><td> 2</td><td>12</td><td> 6.500000</td><td>1</td><td>0</td><td>1</td><td>0.5000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>T_run_4</td><td>MT-ND3</td><td>I6T </td><td>Missense  </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10078_T_C</td><td> 2</td><td> 5</td><td> 3.000000</td><td>1</td><td>0</td><td>1</td><td>0.5000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>T_run_4</td><td>MT-ND3</td><td>L7S </td><td>Missense  </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>1008_A_G </td><td> 2</td><td> 2</td><td> 1.500000</td><td>1</td><td>1</td><td>0</td><td>0.5000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>0.07</td><td>NA</td><td></td><td>NA     </td><td>NA    </td><td>NA  </td><td>NA        </td><td>NA</td><td>A_G</td><td>transition</td></tr>\n",
       "\t<tr><td>10081_T_C</td><td> 3</td><td>33</td><td>11.666667</td><td>2</td><td>0</td><td>1</td><td>0.3333333</td><td>0.0005104645</td><td></td><td>⋯</td><td>0.00</td><td>NA</td><td></td><td>NA     </td><td>MT-ND3</td><td>M8T </td><td>Missense  </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10083_A_G</td><td> 2</td><td> 5</td><td> 3.000000</td><td>1</td><td>0</td><td>1</td><td>0.5000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>0.00</td><td>NA</td><td></td><td>NA     </td><td>MT-ND3</td><td>I9V </td><td>Missense  </td><td>NA</td><td>A_G</td><td>transition</td></tr>\n",
       "\t<tr><td>10084_T_C</td><td> 4</td><td> 8</td><td> 3.250000</td><td>2</td><td>0</td><td>2</td><td>0.5000000</td><td>0.0006806194</td><td></td><td>⋯</td><td>0.83</td><td> 1</td><td></td><td>NA     </td><td>MT-ND3</td><td>I9T </td><td>Missense  </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10088_C_T</td><td> 6</td><td> 7</td><td> 2.000000</td><td>5</td><td>0</td><td>1</td><td>0.1666667</td><td>0.0010209290</td><td></td><td>⋯</td><td>0.04</td><td>NA</td><td></td><td>NA     </td><td>MT-ND3</td><td>N10N</td><td>Synonymous</td><td>NA</td><td>C_T</td><td>transition</td></tr>\n",
       "\t<tr><td>10093_T_C</td><td> 9</td><td> 5</td><td> 1.444444</td><td>8</td><td>0</td><td>1</td><td>0.1111111</td><td>0.0015313936</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-ND3</td><td>L12P</td><td>Missense  </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10098_G_A</td><td> 7</td><td>13</td><td> 3.000000</td><td>5</td><td>0</td><td>2</td><td>0.2857143</td><td>0.0011910839</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-ND3</td><td>A14T</td><td>Missense  </td><td>NA</td><td>G_A</td><td>transition</td></tr>\n",
       "\t<tr><td>10101_T_C</td><td>10</td><td> 4</td><td> 1.500000</td><td>8</td><td>0</td><td>2</td><td>0.2000000</td><td>0.0017015484</td><td></td><td>⋯</td><td>0.25</td><td> 1</td><td></td><td>NA     </td><td>MT-ND3</td><td>L15L</td><td>Synonymous</td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10102_T_C</td><td> 2</td><td> 3</td><td> 2.000000</td><td>1</td><td>0</td><td>1</td><td>0.5000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-ND3</td><td>L15S</td><td>Missense  </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10114_T_C</td><td> 7</td><td>28</td><td>12.285714</td><td>2</td><td>0</td><td>5</td><td>0.7142857</td><td>0.0011910839</td><td></td><td>⋯</td><td>0.01</td><td>NA</td><td></td><td>NA     </td><td>MT-ND3</td><td>I19T</td><td>Missense  </td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10115_T_C</td><td> 4</td><td>18</td><td> 6.250000</td><td>1</td><td>1</td><td>2</td><td>0.7500000</td><td>0.0006806194</td><td></td><td>⋯</td><td>0.03</td><td> 5</td><td></td><td>NA     </td><td>MT-ND3</td><td>I19I</td><td>Synonymous</td><td>NA</td><td>T_C</td><td>transition</td></tr>\n",
       "\t<tr><td>10126_G_A</td><td> 5</td><td> 7</td><td> 2.200000</td><td>4</td><td>0</td><td>1</td><td>0.2000000</td><td>0.0008507742</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-ND3</td><td>W23*</td><td>Nonsense  </td><td>NA</td><td>G_A</td><td>transition</td></tr>\n",
       "\t<tr><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋱</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><td>9914_A_G</td><td>13</td><td>36</td><td>3.692308</td><td>12</td><td>0</td><td>1</td><td>0.07692308</td><td>0.0022120129</td><td></td><td>⋯</td><td>0.00</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>E236E</td><td>Synonymous</td><td>NA                           </td><td>A_G</td><td>transition  </td></tr>\n",
       "\t<tr><td>9915_G_A</td><td>12</td><td>11</td><td>2.000000</td><td> 9</td><td>2</td><td>1</td><td>0.25000000</td><td>0.0020418581</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>A237T</td><td>Missense  </td><td>NA                           </td><td>G_A</td><td>transition  </td></tr>\n",
       "\t<tr><td>9918_G_A</td><td>51</td><td>12</td><td>1.235294</td><td>49</td><td>1</td><td>1</td><td>0.03921569</td><td>0.0086778969</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>A238T</td><td>Missense  </td><td>NA                           </td><td>G_A</td><td>transition  </td></tr>\n",
       "\t<tr><td>9921_G_A</td><td> 7</td><td>11</td><td>2.428571</td><td> 6</td><td>0</td><td>1</td><td>0.14285714</td><td>0.0011910839</td><td></td><td>⋯</td><td>0.13</td><td> 1</td><td></td><td>NA     </td><td>MT-CO3</td><td>A239T</td><td>Missense  </td><td>Possible LHON helper mutation</td><td>G_A</td><td>transition  </td></tr>\n",
       "\t<tr><td>9925_G_A</td><td> 8</td><td> 5</td><td>1.875000</td><td> 6</td><td>0</td><td>2</td><td>0.25000000</td><td>0.0013612387</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>W240*</td><td>Nonsense  </td><td>NA                           </td><td>G_A</td><td>transition  </td></tr>\n",
       "\t<tr><td>9927_T_C</td><td> 4</td><td>19</td><td>6.750000</td><td> 1</td><td>1</td><td>2</td><td>0.75000000</td><td>0.0006806194</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>Y241H</td><td>Missense  </td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "\t<tr><td>9928_A_G</td><td> 6</td><td> 4</td><td>1.500000</td><td> 5</td><td>0</td><td>1</td><td>0.16666667</td><td>0.0010209290</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>Y241C</td><td>Missense  </td><td>NA                           </td><td>A_G</td><td>transition  </td></tr>\n",
       "\t<tr><td>9932_G_A</td><td>15</td><td>32</td><td>3.533333</td><td>12</td><td>0</td><td>3</td><td>0.20000000</td><td>0.0025523226</td><td></td><td>⋯</td><td>0.40</td><td> 2</td><td></td><td>NA     </td><td>MT-CO3</td><td>W242W</td><td>Synonymous</td><td>NA                           </td><td>G_A</td><td>transition  </td></tr>\n",
       "\t<tr><td>9934_A_G</td><td> 2</td><td> 3</td><td>2.000000</td><td> 1</td><td>0</td><td>1</td><td>0.50000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>T_run_4</td><td>MT-CO3</td><td>H243R</td><td>Missense  </td><td>NA                           </td><td>A_G</td><td>transition  </td></tr>\n",
       "\t<tr><td>9938_T_C</td><td> 4</td><td> 5</td><td>2.250000</td><td> 2</td><td>1</td><td>1</td><td>0.50000000</td><td>0.0006806194</td><td></td><td>⋯</td><td>0.13</td><td>NA</td><td></td><td>T_run_4</td><td>MT-CO3</td><td>F244F</td><td>Synonymous</td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "\t<tr><td>9939_G_A</td><td>10</td><td>10</td><td>2.800000</td><td> 7</td><td>0</td><td>3</td><td>0.30000000</td><td>0.0017015484</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>V245M</td><td>Missense  </td><td>NA                           </td><td>G_A</td><td>transition  </td></tr>\n",
       "\t<tr><td>9940_T_C</td><td> 2</td><td>12</td><td>7.000000</td><td> 0</td><td>1</td><td>1</td><td>1.00000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>V245A</td><td>Missense  </td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "\t<tr><td>9942_G_A</td><td>15</td><td>36</td><td>6.800000</td><td> 8</td><td>0</td><td>7</td><td>0.46666667</td><td>0.0025523226</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>D246N</td><td>Missense  </td><td>NA                           </td><td>G_A</td><td>transition  </td></tr>\n",
       "\t<tr><td>9947_G_A</td><td>10</td><td>32</td><td>4.500000</td><td> 6</td><td>2</td><td>2</td><td>0.40000000</td><td>0.0017015484</td><td></td><td>⋯</td><td>1.13</td><td> 2</td><td></td><td>NA     </td><td>MT-CO3</td><td>V247V</td><td>Synonymous</td><td>NA                           </td><td>G_A</td><td>transition  </td></tr>\n",
       "\t<tr><td>9948_G_A</td><td>10</td><td>17</td><td>5.000000</td><td> 6</td><td>0</td><td>4</td><td>0.40000000</td><td>0.0017015484</td><td></td><td>⋯</td><td>0.21</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>V248I</td><td>Missense  </td><td>NA                           </td><td>G_A</td><td>transition  </td></tr>\n",
       "\t<tr><td>9948_G_C</td><td> 2</td><td>10</td><td>5.500000</td><td> 1</td><td>0</td><td>1</td><td>0.50000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>V248L</td><td>Missense  </td><td>NA                           </td><td>G_C</td><td>transversion</td></tr>\n",
       "\t<tr><td>9949_T_C</td><td> 2</td><td> 7</td><td>4.000000</td><td> 1</td><td>0</td><td>1</td><td>0.50000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>V248A</td><td>Missense  </td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "\t<tr><td>9950_T_C</td><td>13</td><td> 6</td><td>1.384615</td><td>12</td><td>0</td><td>1</td><td>0.07692308</td><td>0.0022120129</td><td></td><td>⋯</td><td>4.73</td><td>17</td><td></td><td>NA     </td><td>MT-CO3</td><td>V248V</td><td>Synonymous</td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "\t<tr><td>9964_A_G</td><td> 6</td><td>13</td><td>5.166667</td><td> 3</td><td>0</td><td>3</td><td>0.50000000</td><td>0.0010209290</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>Y253C</td><td>Missense  </td><td>NA                           </td><td>A_G</td><td>transition  </td></tr>\n",
       "\t<tr><td>9965_T_C</td><td> 3</td><td> 4</td><td>2.000000</td><td> 2</td><td>0</td><td>1</td><td>0.33333333</td><td>0.0005104645</td><td></td><td>⋯</td><td>0.01</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>Y253Y</td><td>Synonymous</td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "\t<tr><td>9966_G_A</td><td> 6</td><td> 8</td><td>2.833333</td><td> 4</td><td>0</td><td>2</td><td>0.33333333</td><td>0.0010209290</td><td></td><td>⋯</td><td>0.57</td><td> 5</td><td></td><td>NA     </td><td>MT-CO3</td><td>V254I</td><td>Missense  </td><td>LHON possible helper variant </td><td>G_A</td><td>transition  </td></tr>\n",
       "\t<tr><td>9967_T_C</td><td> 4</td><td> 3</td><td>1.500000</td><td> 3</td><td>0</td><td>1</td><td>0.25000000</td><td>0.0006806194</td><td></td><td>⋯</td><td>0.01</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>V254A</td><td>Missense  </td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "\t<tr><td>9973_T_C</td><td> 7</td><td>18</td><td>4.000000</td><td> 5</td><td>0</td><td>2</td><td>0.28571429</td><td>0.0011910839</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>I256T</td><td>Missense  </td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "\t<tr><td>9977_T_C</td><td> 8</td><td> 4</td><td>1.750000</td><td> 6</td><td>0</td><td>2</td><td>0.25000000</td><td>0.0013612387</td><td></td><td>⋯</td><td>0.03</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>Y257Y</td><td>Synonymous</td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "\t<tr><td>9978_T_C</td><td> 4</td><td>21</td><td>7.500000</td><td> 1</td><td>1</td><td>2</td><td>0.75000000</td><td>0.0006806194</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>W258R</td><td>Missense  </td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "\t<tr><td>9985_G_A</td><td>14</td><td> 3</td><td>1.142857</td><td>13</td><td>0</td><td>1</td><td>0.07142857</td><td>0.0023821678</td><td></td><td>⋯</td><td>0.00</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>G260E</td><td>Missense  </td><td>NA                           </td><td>G_A</td><td>transition  </td></tr>\n",
       "\t<tr><td>9986_G_A</td><td>10</td><td> 7</td><td>1.600000</td><td> 9</td><td>0</td><td>1</td><td>0.10000000</td><td>0.0017015484</td><td></td><td>⋯</td><td>0.21</td><td> 1</td><td></td><td>NA     </td><td>MT-CO3</td><td>G260G</td><td>Synonymous</td><td>NA                           </td><td>G_A</td><td>transition  </td></tr>\n",
       "\t<tr><td>9987_T_C</td><td> 5</td><td>11</td><td>3.200000</td><td> 3</td><td>1</td><td>1</td><td>0.40000000</td><td>0.0008507742</td><td></td><td>⋯</td><td>0.00</td><td>NA</td><td></td><td>NA     </td><td>MT-CO3</td><td>S261P</td><td>Missense  </td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "\t<tr><td>9993_T_C</td><td> 4</td><td>10</td><td>3.750000</td><td> 2</td><td>0</td><td>2</td><td>0.50000000</td><td>0.0006806194</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>NA     </td><td>NA    </td><td>NA   </td><td>NA        </td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "\t<tr><td>9998_T_C</td><td> 2</td><td> 3</td><td>2.000000</td><td> 1</td><td>0</td><td>1</td><td>0.50000000</td><td>0.0003403097</td><td></td><td>⋯</td><td>  NA</td><td>NA</td><td></td><td>T_run_4</td><td>NA    </td><td>NA   </td><td>NA        </td><td>NA                           </td><td>T_C</td><td>transition  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3782 × 25\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " Variants & CellN & max\\_count & mean\\_count & ct1\\_cellN & ct2\\_cellN & ct3plus\\_cellN & pct\\_ct2plus & cellNPCT & hyper\\_label & ⋯ & freq\\_eurasian & n\\_haplos & blacklist\\_region & homopolymer & gene & aachange & impact & Disease & changes & types\\\\\n",
       " <chr> & <int> & <dbl> & <dbl> & <int> & <int> & <int> & <dbl> & <dbl> & <chr> & ⋯ & <dbl> & <int> & <chr> & <chr> & <chr> & <chr{[}1d{]}> & <chr{[}1d{]}> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t 1000\\_T\\_C  &  3 &  4 &  2.000000 & 2 & 0 & 1 & 0.3333333 & 0.0005104645 &  & ⋯ &   NA & NA &  & NA      & NA     & NA   & NA         & NA & T\\_C & transition\\\\\n",
       "\t 10005\\_A\\_G &  2 &  3 &  2.000000 & 1 & 0 & 1 & 0.5000000 & 0.0003403097 &  & ⋯ & 0.02 & NA &  & NA      & NA     & NA   & NA         & NA & A\\_G & transition\\\\\n",
       "\t 10007\\_T\\_C &  2 &  3 &  2.500000 & 0 & 1 & 1 & 1.0000000 & 0.0003403097 &  & ⋯ & 0.10 & NA &  & NA      & NA     & NA   & NA         & NA & T\\_C & transition\\\\\n",
       "\t 10009\\_G\\_A &  2 & 10 &  5.500000 & 1 & 0 & 1 & 0.5000000 & 0.0003403097 &  & ⋯ &   NA & NA &  & NA      & NA     & NA   & NA         & NA & G\\_A & transition\\\\\n",
       "\t 10010\\_T\\_C &  3 & 12 &  4.666667 & 2 & 0 & 1 & 0.3333333 & 0.0005104645 &  & ⋯ &   NA & NA &  & NA      & NA     & NA   & NA         & NA & T\\_C & transition\\\\\n",
       "\t 10020\\_T\\_C &  6 & 13 &  4.833333 & 1 & 2 & 3 & 0.8333333 & 0.0010209290 &  & ⋯ & 0.01 & NA &  & NA      & NA     & NA   & NA         & NA & T\\_C & transition\\\\\n",
       "\t 10031\\_T\\_C &  3 &  2 &  1.333333 & 2 & 1 & 0 & 0.3333333 & 0.0005104645 &  & ⋯ & 0.18 &  2 &  & NA      & NA     & NA   & NA         & NA & T\\_C & transition\\\\\n",
       "\t 1004\\_G\\_A  &  4 &  3 &  2.000000 & 2 & 0 & 2 & 0.5000000 & 0.0006806194 &  & ⋯ &   NA & NA &  & NA      & NA     & NA   & NA         & NA & G\\_A & transition\\\\\n",
       "\t 10045\\_T\\_C &  2 & 32 & 16.500000 & 1 & 0 & 1 & 0.5000000 & 0.0003403097 &  & ⋯ & 0.04 & NA &  & NA      & NA     & NA   & NA         & NA & T\\_C & transition\\\\\n",
       "\t 1005\\_T\\_C  &  6 &  7 &  2.000000 & 5 & 0 & 1 & 0.1666667 & 0.0010209290 &  & ⋯ & 0.62 &  8 &  & NA      & NA     & NA   & NA         & NA & T\\_C & transition\\\\\n",
       "\t 10054\\_G\\_A &  7 & 11 &  2.428571 & 6 & 0 & 1 & 0.1428571 & 0.0011910839 &  & ⋯ &   NA & NA &  & NA      & NA     & NA   & NA         & NA & G\\_A & transition\\\\\n",
       "\t 1006\\_T\\_C  &  9 & 30 &  4.555556 & 7 & 0 & 2 & 0.2222222 & 0.0015313936 &  & ⋯ & 0.00 & NA &  & NA      & NA     & NA   & NA         & NA & T\\_C & transition\\\\\n",
       "\t 10063\\_A\\_G &  4 &  4 &  2.750000 & 1 & 1 & 2 & 0.7500000 & 0.0006806194 &  & ⋯ &   NA & NA &  & NA      & MT-ND3 & N2S  & Missense   & NA & A\\_G & transition\\\\\n",
       "\t 10066\\_T\\_C &  4 &  5 &  2.000000 & 3 & 0 & 1 & 0.2500000 & 0.0006806194 &  & ⋯ &   NA & NA &  & NA      & MT-ND3 & F3S  & Missense   & NA & T\\_C & transition\\\\\n",
       "\t 10068\\_G\\_A & 10 &  4 &  1.900000 & 6 & 1 & 3 & 0.4000000 & 0.0017015484 &  & ⋯ &   NA & NA &  & NA      & MT-ND3 & A4T  & Missense   & NA & G\\_A & transition\\\\\n",
       "\t 10071\\_T\\_C &  4 &  2 &  1.250000 & 3 & 1 & 0 & 0.2500000 & 0.0006806194 &  & ⋯ & 0.00 & NA &  & NA      & MT-ND3 & L5L  & Synonymous & NA & T\\_C & transition\\\\\n",
       "\t 10075\\_T\\_C &  2 & 12 &  6.500000 & 1 & 0 & 1 & 0.5000000 & 0.0003403097 &  & ⋯ &   NA & NA &  & T\\_run\\_4 & MT-ND3 & I6T  & Missense   & NA & T\\_C & transition\\\\\n",
       "\t 10078\\_T\\_C &  2 &  5 &  3.000000 & 1 & 0 & 1 & 0.5000000 & 0.0003403097 &  & ⋯ &   NA & NA &  & T\\_run\\_4 & MT-ND3 & L7S  & Missense   & NA & T\\_C & transition\\\\\n",
       "\t 1008\\_A\\_G  &  2 &  2 &  1.500000 & 1 & 1 & 0 & 0.5000000 & 0.0003403097 &  & ⋯ & 0.07 & NA &  & NA      & NA     & NA   & NA         & NA & A\\_G & transition\\\\\n",
       "\t 10081\\_T\\_C &  3 & 33 & 11.666667 & 2 & 0 & 1 & 0.3333333 & 0.0005104645 &  & ⋯ & 0.00 & NA &  & NA      & MT-ND3 & M8T  & Missense   & NA & T\\_C & transition\\\\\n",
       "\t 10083\\_A\\_G &  2 &  5 &  3.000000 & 1 & 0 & 1 & 0.5000000 & 0.0003403097 &  & ⋯ & 0.00 & NA &  & NA      & MT-ND3 & I9V  & Missense   & NA & A\\_G & transition\\\\\n",
       "\t 10084\\_T\\_C &  4 &  8 &  3.250000 & 2 & 0 & 2 & 0.5000000 & 0.0006806194 &  & ⋯ & 0.83 &  1 &  & NA      & MT-ND3 & I9T  & Missense   & NA & T\\_C & transition\\\\\n",
       "\t 10088\\_C\\_T &  6 &  7 &  2.000000 & 5 & 0 & 1 & 0.1666667 & 0.0010209290 &  & ⋯ & 0.04 & NA &  & NA      & MT-ND3 & N10N & Synonymous & NA & C\\_T & transition\\\\\n",
       "\t 10093\\_T\\_C &  9 &  5 &  1.444444 & 8 & 0 & 1 & 0.1111111 & 0.0015313936 &  & ⋯ &   NA & NA &  & NA      & MT-ND3 & L12P & Missense   & NA & T\\_C & transition\\\\\n",
       "\t 10098\\_G\\_A &  7 & 13 &  3.000000 & 5 & 0 & 2 & 0.2857143 & 0.0011910839 &  & ⋯ &   NA & NA &  & NA      & MT-ND3 & A14T & Missense   & NA & G\\_A & transition\\\\\n",
       "\t 10101\\_T\\_C & 10 &  4 &  1.500000 & 8 & 0 & 2 & 0.2000000 & 0.0017015484 &  & ⋯ & 0.25 &  1 &  & NA      & MT-ND3 & L15L & Synonymous & NA & T\\_C & transition\\\\\n",
       "\t 10102\\_T\\_C &  2 &  3 &  2.000000 & 1 & 0 & 1 & 0.5000000 & 0.0003403097 &  & ⋯ &   NA & NA &  & NA      & MT-ND3 & L15S & Missense   & NA & T\\_C & transition\\\\\n",
       "\t 10114\\_T\\_C &  7 & 28 & 12.285714 & 2 & 0 & 5 & 0.7142857 & 0.0011910839 &  & ⋯ & 0.01 & NA &  & NA      & MT-ND3 & I19T & Missense   & NA & T\\_C & transition\\\\\n",
       "\t 10115\\_T\\_C &  4 & 18 &  6.250000 & 1 & 1 & 2 & 0.7500000 & 0.0006806194 &  & ⋯ & 0.03 &  5 &  & NA      & MT-ND3 & I19I & Synonymous & NA & T\\_C & transition\\\\\n",
       "\t 10126\\_G\\_A &  5 &  7 &  2.200000 & 4 & 0 & 1 & 0.2000000 & 0.0008507742 &  & ⋯ &   NA & NA &  & NA      & MT-ND3 & W23* & Nonsense   & NA & G\\_A & transition\\\\\n",
       "\t ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋱ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\t 9914\\_A\\_G & 13 & 36 & 3.692308 & 12 & 0 & 1 & 0.07692308 & 0.0022120129 &  & ⋯ & 0.00 & NA &  & NA      & MT-CO3 & E236E & Synonymous & NA                            & A\\_G & transition  \\\\\n",
       "\t 9915\\_G\\_A & 12 & 11 & 2.000000 &  9 & 2 & 1 & 0.25000000 & 0.0020418581 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & A237T & Missense   & NA                            & G\\_A & transition  \\\\\n",
       "\t 9918\\_G\\_A & 51 & 12 & 1.235294 & 49 & 1 & 1 & 0.03921569 & 0.0086778969 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & A238T & Missense   & NA                            & G\\_A & transition  \\\\\n",
       "\t 9921\\_G\\_A &  7 & 11 & 2.428571 &  6 & 0 & 1 & 0.14285714 & 0.0011910839 &  & ⋯ & 0.13 &  1 &  & NA      & MT-CO3 & A239T & Missense   & Possible LHON helper mutation & G\\_A & transition  \\\\\n",
       "\t 9925\\_G\\_A &  8 &  5 & 1.875000 &  6 & 0 & 2 & 0.25000000 & 0.0013612387 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & W240* & Nonsense   & NA                            & G\\_A & transition  \\\\\n",
       "\t 9927\\_T\\_C &  4 & 19 & 6.750000 &  1 & 1 & 2 & 0.75000000 & 0.0006806194 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & Y241H & Missense   & NA                            & T\\_C & transition  \\\\\n",
       "\t 9928\\_A\\_G &  6 &  4 & 1.500000 &  5 & 0 & 1 & 0.16666667 & 0.0010209290 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & Y241C & Missense   & NA                            & A\\_G & transition  \\\\\n",
       "\t 9932\\_G\\_A & 15 & 32 & 3.533333 & 12 & 0 & 3 & 0.20000000 & 0.0025523226 &  & ⋯ & 0.40 &  2 &  & NA      & MT-CO3 & W242W & Synonymous & NA                            & G\\_A & transition  \\\\\n",
       "\t 9934\\_A\\_G &  2 &  3 & 2.000000 &  1 & 0 & 1 & 0.50000000 & 0.0003403097 &  & ⋯ &   NA & NA &  & T\\_run\\_4 & MT-CO3 & H243R & Missense   & NA                            & A\\_G & transition  \\\\\n",
       "\t 9938\\_T\\_C &  4 &  5 & 2.250000 &  2 & 1 & 1 & 0.50000000 & 0.0006806194 &  & ⋯ & 0.13 & NA &  & T\\_run\\_4 & MT-CO3 & F244F & Synonymous & NA                            & T\\_C & transition  \\\\\n",
       "\t 9939\\_G\\_A & 10 & 10 & 2.800000 &  7 & 0 & 3 & 0.30000000 & 0.0017015484 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & V245M & Missense   & NA                            & G\\_A & transition  \\\\\n",
       "\t 9940\\_T\\_C &  2 & 12 & 7.000000 &  0 & 1 & 1 & 1.00000000 & 0.0003403097 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & V245A & Missense   & NA                            & T\\_C & transition  \\\\\n",
       "\t 9942\\_G\\_A & 15 & 36 & 6.800000 &  8 & 0 & 7 & 0.46666667 & 0.0025523226 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & D246N & Missense   & NA                            & G\\_A & transition  \\\\\n",
       "\t 9947\\_G\\_A & 10 & 32 & 4.500000 &  6 & 2 & 2 & 0.40000000 & 0.0017015484 &  & ⋯ & 1.13 &  2 &  & NA      & MT-CO3 & V247V & Synonymous & NA                            & G\\_A & transition  \\\\\n",
       "\t 9948\\_G\\_A & 10 & 17 & 5.000000 &  6 & 0 & 4 & 0.40000000 & 0.0017015484 &  & ⋯ & 0.21 & NA &  & NA      & MT-CO3 & V248I & Missense   & NA                            & G\\_A & transition  \\\\\n",
       "\t 9948\\_G\\_C &  2 & 10 & 5.500000 &  1 & 0 & 1 & 0.50000000 & 0.0003403097 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & V248L & Missense   & NA                            & G\\_C & transversion\\\\\n",
       "\t 9949\\_T\\_C &  2 &  7 & 4.000000 &  1 & 0 & 1 & 0.50000000 & 0.0003403097 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & V248A & Missense   & NA                            & T\\_C & transition  \\\\\n",
       "\t 9950\\_T\\_C & 13 &  6 & 1.384615 & 12 & 0 & 1 & 0.07692308 & 0.0022120129 &  & ⋯ & 4.73 & 17 &  & NA      & MT-CO3 & V248V & Synonymous & NA                            & T\\_C & transition  \\\\\n",
       "\t 9964\\_A\\_G &  6 & 13 & 5.166667 &  3 & 0 & 3 & 0.50000000 & 0.0010209290 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & Y253C & Missense   & NA                            & A\\_G & transition  \\\\\n",
       "\t 9965\\_T\\_C &  3 &  4 & 2.000000 &  2 & 0 & 1 & 0.33333333 & 0.0005104645 &  & ⋯ & 0.01 & NA &  & NA      & MT-CO3 & Y253Y & Synonymous & NA                            & T\\_C & transition  \\\\\n",
       "\t 9966\\_G\\_A &  6 &  8 & 2.833333 &  4 & 0 & 2 & 0.33333333 & 0.0010209290 &  & ⋯ & 0.57 &  5 &  & NA      & MT-CO3 & V254I & Missense   & LHON possible helper variant  & G\\_A & transition  \\\\\n",
       "\t 9967\\_T\\_C &  4 &  3 & 1.500000 &  3 & 0 & 1 & 0.25000000 & 0.0006806194 &  & ⋯ & 0.01 & NA &  & NA      & MT-CO3 & V254A & Missense   & NA                            & T\\_C & transition  \\\\\n",
       "\t 9973\\_T\\_C &  7 & 18 & 4.000000 &  5 & 0 & 2 & 0.28571429 & 0.0011910839 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & I256T & Missense   & NA                            & T\\_C & transition  \\\\\n",
       "\t 9977\\_T\\_C &  8 &  4 & 1.750000 &  6 & 0 & 2 & 0.25000000 & 0.0013612387 &  & ⋯ & 0.03 & NA &  & NA      & MT-CO3 & Y257Y & Synonymous & NA                            & T\\_C & transition  \\\\\n",
       "\t 9978\\_T\\_C &  4 & 21 & 7.500000 &  1 & 1 & 2 & 0.75000000 & 0.0006806194 &  & ⋯ &   NA & NA &  & NA      & MT-CO3 & W258R & Missense   & NA                            & T\\_C & transition  \\\\\n",
       "\t 9985\\_G\\_A & 14 &  3 & 1.142857 & 13 & 0 & 1 & 0.07142857 & 0.0023821678 &  & ⋯ & 0.00 & NA &  & NA      & MT-CO3 & G260E & Missense   & NA                            & G\\_A & transition  \\\\\n",
       "\t 9986\\_G\\_A & 10 &  7 & 1.600000 &  9 & 0 & 1 & 0.10000000 & 0.0017015484 &  & ⋯ & 0.21 &  1 &  & NA      & MT-CO3 & G260G & Synonymous & NA                            & G\\_A & transition  \\\\\n",
       "\t 9987\\_T\\_C &  5 & 11 & 3.200000 &  3 & 1 & 1 & 0.40000000 & 0.0008507742 &  & ⋯ & 0.00 & NA &  & NA      & MT-CO3 & S261P & Missense   & NA                            & T\\_C & transition  \\\\\n",
       "\t 9993\\_T\\_C &  4 & 10 & 3.750000 &  2 & 0 & 2 & 0.50000000 & 0.0006806194 &  & ⋯ &   NA & NA &  & NA      & NA     & NA    & NA         & NA                            & T\\_C & transition  \\\\\n",
       "\t 9998\\_T\\_C &  2 &  3 & 2.000000 &  1 & 0 & 1 & 0.50000000 & 0.0003403097 &  & ⋯ &   NA & NA &  & T\\_run\\_4 & NA     & NA    & NA         & NA                            & T\\_C & transition  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3782 × 25\n",
       "\n",
       "| Variants &lt;chr&gt; | CellN &lt;int&gt; | max_count &lt;dbl&gt; | mean_count &lt;dbl&gt; | ct1_cellN &lt;int&gt; | ct2_cellN &lt;int&gt; | ct3plus_cellN &lt;int&gt; | pct_ct2plus &lt;dbl&gt; | cellNPCT &lt;dbl&gt; | hyper_label &lt;chr&gt; | ⋯ ⋯ | freq_eurasian &lt;dbl&gt; | n_haplos &lt;int&gt; | blacklist_region &lt;chr&gt; | homopolymer &lt;chr&gt; | gene &lt;chr&gt; | aachange &lt;chr[1d]&gt; | impact &lt;chr[1d]&gt; | Disease &lt;chr&gt; | changes &lt;chr&gt; | types &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1000_T_C  |  3 |  4 |  2.000000 | 2 | 0 | 1 | 0.3333333 | 0.0005104645 | <!----> | ⋯ |   NA | NA | <!----> | NA      | NA     | NA   | NA         | NA | T_C | transition |\n",
       "| 10005_A_G |  2 |  3 |  2.000000 | 1 | 0 | 1 | 0.5000000 | 0.0003403097 | <!----> | ⋯ | 0.02 | NA | <!----> | NA      | NA     | NA   | NA         | NA | A_G | transition |\n",
       "| 10007_T_C |  2 |  3 |  2.500000 | 0 | 1 | 1 | 1.0000000 | 0.0003403097 | <!----> | ⋯ | 0.10 | NA | <!----> | NA      | NA     | NA   | NA         | NA | T_C | transition |\n",
       "| 10009_G_A |  2 | 10 |  5.500000 | 1 | 0 | 1 | 0.5000000 | 0.0003403097 | <!----> | ⋯ |   NA | NA | <!----> | NA      | NA     | NA   | NA         | NA | G_A | transition |\n",
       "| 10010_T_C |  3 | 12 |  4.666667 | 2 | 0 | 1 | 0.3333333 | 0.0005104645 | <!----> | ⋯ |   NA | NA | <!----> | NA      | NA     | NA   | NA         | NA | T_C | transition |\n",
       "| 10020_T_C |  6 | 13 |  4.833333 | 1 | 2 | 3 | 0.8333333 | 0.0010209290 | <!----> | ⋯ | 0.01 | NA | <!----> | NA      | NA     | NA   | NA         | NA | T_C | transition |\n",
       "| 10031_T_C |  3 |  2 |  1.333333 | 2 | 1 | 0 | 0.3333333 | 0.0005104645 | <!----> | ⋯ | 0.18 |  2 | <!----> | NA      | NA     | NA   | NA         | NA | T_C | transition |\n",
       "| 1004_G_A  |  4 |  3 |  2.000000 | 2 | 0 | 2 | 0.5000000 | 0.0006806194 | <!----> | ⋯ |   NA | NA | <!----> | NA      | NA     | NA   | NA         | NA | G_A | transition |\n",
       "| 10045_T_C |  2 | 32 | 16.500000 | 1 | 0 | 1 | 0.5000000 | 0.0003403097 | <!----> | ⋯ | 0.04 | NA | <!----> | NA      | NA     | NA   | NA         | NA | T_C | transition |\n",
       "| 1005_T_C  |  6 |  7 |  2.000000 | 5 | 0 | 1 | 0.1666667 | 0.0010209290 | <!----> | ⋯ | 0.62 |  8 | <!----> | NA      | NA     | NA   | NA         | NA | T_C | transition |\n",
       "| 10054_G_A |  7 | 11 |  2.428571 | 6 | 0 | 1 | 0.1428571 | 0.0011910839 | <!----> | ⋯ |   NA | NA | <!----> | NA      | NA     | NA   | NA         | NA | G_A | transition |\n",
       "| 1006_T_C  |  9 | 30 |  4.555556 | 7 | 0 | 2 | 0.2222222 | 0.0015313936 | <!----> | ⋯ | 0.00 | NA | <!----> | NA      | NA     | NA   | NA         | NA | T_C | transition |\n",
       "| 10063_A_G |  4 |  4 |  2.750000 | 1 | 1 | 2 | 0.7500000 | 0.0006806194 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-ND3 | N2S  | Missense   | NA | A_G | transition |\n",
       "| 10066_T_C |  4 |  5 |  2.000000 | 3 | 0 | 1 | 0.2500000 | 0.0006806194 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-ND3 | F3S  | Missense   | NA | T_C | transition |\n",
       "| 10068_G_A | 10 |  4 |  1.900000 | 6 | 1 | 3 | 0.4000000 | 0.0017015484 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-ND3 | A4T  | Missense   | NA | G_A | transition |\n",
       "| 10071_T_C |  4 |  2 |  1.250000 | 3 | 1 | 0 | 0.2500000 | 0.0006806194 | <!----> | ⋯ | 0.00 | NA | <!----> | NA      | MT-ND3 | L5L  | Synonymous | NA | T_C | transition |\n",
       "| 10075_T_C |  2 | 12 |  6.500000 | 1 | 0 | 1 | 0.5000000 | 0.0003403097 | <!----> | ⋯ |   NA | NA | <!----> | T_run_4 | MT-ND3 | I6T  | Missense   | NA | T_C | transition |\n",
       "| 10078_T_C |  2 |  5 |  3.000000 | 1 | 0 | 1 | 0.5000000 | 0.0003403097 | <!----> | ⋯ |   NA | NA | <!----> | T_run_4 | MT-ND3 | L7S  | Missense   | NA | T_C | transition |\n",
       "| 1008_A_G  |  2 |  2 |  1.500000 | 1 | 1 | 0 | 0.5000000 | 0.0003403097 | <!----> | ⋯ | 0.07 | NA | <!----> | NA      | NA     | NA   | NA         | NA | A_G | transition |\n",
       "| 10081_T_C |  3 | 33 | 11.666667 | 2 | 0 | 1 | 0.3333333 | 0.0005104645 | <!----> | ⋯ | 0.00 | NA | <!----> | NA      | MT-ND3 | M8T  | Missense   | NA | T_C | transition |\n",
       "| 10083_A_G |  2 |  5 |  3.000000 | 1 | 0 | 1 | 0.5000000 | 0.0003403097 | <!----> | ⋯ | 0.00 | NA | <!----> | NA      | MT-ND3 | I9V  | Missense   | NA | A_G | transition |\n",
       "| 10084_T_C |  4 |  8 |  3.250000 | 2 | 0 | 2 | 0.5000000 | 0.0006806194 | <!----> | ⋯ | 0.83 |  1 | <!----> | NA      | MT-ND3 | I9T  | Missense   | NA | T_C | transition |\n",
       "| 10088_C_T |  6 |  7 |  2.000000 | 5 | 0 | 1 | 0.1666667 | 0.0010209290 | <!----> | ⋯ | 0.04 | NA | <!----> | NA      | MT-ND3 | N10N | Synonymous | NA | C_T | transition |\n",
       "| 10093_T_C |  9 |  5 |  1.444444 | 8 | 0 | 1 | 0.1111111 | 0.0015313936 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-ND3 | L12P | Missense   | NA | T_C | transition |\n",
       "| 10098_G_A |  7 | 13 |  3.000000 | 5 | 0 | 2 | 0.2857143 | 0.0011910839 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-ND3 | A14T | Missense   | NA | G_A | transition |\n",
       "| 10101_T_C | 10 |  4 |  1.500000 | 8 | 0 | 2 | 0.2000000 | 0.0017015484 | <!----> | ⋯ | 0.25 |  1 | <!----> | NA      | MT-ND3 | L15L | Synonymous | NA | T_C | transition |\n",
       "| 10102_T_C |  2 |  3 |  2.000000 | 1 | 0 | 1 | 0.5000000 | 0.0003403097 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-ND3 | L15S | Missense   | NA | T_C | transition |\n",
       "| 10114_T_C |  7 | 28 | 12.285714 | 2 | 0 | 5 | 0.7142857 | 0.0011910839 | <!----> | ⋯ | 0.01 | NA | <!----> | NA      | MT-ND3 | I19T | Missense   | NA | T_C | transition |\n",
       "| 10115_T_C |  4 | 18 |  6.250000 | 1 | 1 | 2 | 0.7500000 | 0.0006806194 | <!----> | ⋯ | 0.03 |  5 | <!----> | NA      | MT-ND3 | I19I | Synonymous | NA | T_C | transition |\n",
       "| 10126_G_A |  5 |  7 |  2.200000 | 4 | 0 | 1 | 0.2000000 | 0.0008507742 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-ND3 | W23* | Nonsense   | NA | G_A | transition |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋱ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| 9914_A_G | 13 | 36 | 3.692308 | 12 | 0 | 1 | 0.07692308 | 0.0022120129 | <!----> | ⋯ | 0.00 | NA | <!----> | NA      | MT-CO3 | E236E | Synonymous | NA                            | A_G | transition   |\n",
       "| 9915_G_A | 12 | 11 | 2.000000 |  9 | 2 | 1 | 0.25000000 | 0.0020418581 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | A237T | Missense   | NA                            | G_A | transition   |\n",
       "| 9918_G_A | 51 | 12 | 1.235294 | 49 | 1 | 1 | 0.03921569 | 0.0086778969 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | A238T | Missense   | NA                            | G_A | transition   |\n",
       "| 9921_G_A |  7 | 11 | 2.428571 |  6 | 0 | 1 | 0.14285714 | 0.0011910839 | <!----> | ⋯ | 0.13 |  1 | <!----> | NA      | MT-CO3 | A239T | Missense   | Possible LHON helper mutation | G_A | transition   |\n",
       "| 9925_G_A |  8 |  5 | 1.875000 |  6 | 0 | 2 | 0.25000000 | 0.0013612387 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | W240* | Nonsense   | NA                            | G_A | transition   |\n",
       "| 9927_T_C |  4 | 19 | 6.750000 |  1 | 1 | 2 | 0.75000000 | 0.0006806194 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | Y241H | Missense   | NA                            | T_C | transition   |\n",
       "| 9928_A_G |  6 |  4 | 1.500000 |  5 | 0 | 1 | 0.16666667 | 0.0010209290 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | Y241C | Missense   | NA                            | A_G | transition   |\n",
       "| 9932_G_A | 15 | 32 | 3.533333 | 12 | 0 | 3 | 0.20000000 | 0.0025523226 | <!----> | ⋯ | 0.40 |  2 | <!----> | NA      | MT-CO3 | W242W | Synonymous | NA                            | G_A | transition   |\n",
       "| 9934_A_G |  2 |  3 | 2.000000 |  1 | 0 | 1 | 0.50000000 | 0.0003403097 | <!----> | ⋯ |   NA | NA | <!----> | T_run_4 | MT-CO3 | H243R | Missense   | NA                            | A_G | transition   |\n",
       "| 9938_T_C |  4 |  5 | 2.250000 |  2 | 1 | 1 | 0.50000000 | 0.0006806194 | <!----> | ⋯ | 0.13 | NA | <!----> | T_run_4 | MT-CO3 | F244F | Synonymous | NA                            | T_C | transition   |\n",
       "| 9939_G_A | 10 | 10 | 2.800000 |  7 | 0 | 3 | 0.30000000 | 0.0017015484 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | V245M | Missense   | NA                            | G_A | transition   |\n",
       "| 9940_T_C |  2 | 12 | 7.000000 |  0 | 1 | 1 | 1.00000000 | 0.0003403097 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | V245A | Missense   | NA                            | T_C | transition   |\n",
       "| 9942_G_A | 15 | 36 | 6.800000 |  8 | 0 | 7 | 0.46666667 | 0.0025523226 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | D246N | Missense   | NA                            | G_A | transition   |\n",
       "| 9947_G_A | 10 | 32 | 4.500000 |  6 | 2 | 2 | 0.40000000 | 0.0017015484 | <!----> | ⋯ | 1.13 |  2 | <!----> | NA      | MT-CO3 | V247V | Synonymous | NA                            | G_A | transition   |\n",
       "| 9948_G_A | 10 | 17 | 5.000000 |  6 | 0 | 4 | 0.40000000 | 0.0017015484 | <!----> | ⋯ | 0.21 | NA | <!----> | NA      | MT-CO3 | V248I | Missense   | NA                            | G_A | transition   |\n",
       "| 9948_G_C |  2 | 10 | 5.500000 |  1 | 0 | 1 | 0.50000000 | 0.0003403097 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | V248L | Missense   | NA                            | G_C | transversion |\n",
       "| 9949_T_C |  2 |  7 | 4.000000 |  1 | 0 | 1 | 0.50000000 | 0.0003403097 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | V248A | Missense   | NA                            | T_C | transition   |\n",
       "| 9950_T_C | 13 |  6 | 1.384615 | 12 | 0 | 1 | 0.07692308 | 0.0022120129 | <!----> | ⋯ | 4.73 | 17 | <!----> | NA      | MT-CO3 | V248V | Synonymous | NA                            | T_C | transition   |\n",
       "| 9964_A_G |  6 | 13 | 5.166667 |  3 | 0 | 3 | 0.50000000 | 0.0010209290 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | Y253C | Missense   | NA                            | A_G | transition   |\n",
       "| 9965_T_C |  3 |  4 | 2.000000 |  2 | 0 | 1 | 0.33333333 | 0.0005104645 | <!----> | ⋯ | 0.01 | NA | <!----> | NA      | MT-CO3 | Y253Y | Synonymous | NA                            | T_C | transition   |\n",
       "| 9966_G_A |  6 |  8 | 2.833333 |  4 | 0 | 2 | 0.33333333 | 0.0010209290 | <!----> | ⋯ | 0.57 |  5 | <!----> | NA      | MT-CO3 | V254I | Missense   | LHON possible helper variant  | G_A | transition   |\n",
       "| 9967_T_C |  4 |  3 | 1.500000 |  3 | 0 | 1 | 0.25000000 | 0.0006806194 | <!----> | ⋯ | 0.01 | NA | <!----> | NA      | MT-CO3 | V254A | Missense   | NA                            | T_C | transition   |\n",
       "| 9973_T_C |  7 | 18 | 4.000000 |  5 | 0 | 2 | 0.28571429 | 0.0011910839 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | I256T | Missense   | NA                            | T_C | transition   |\n",
       "| 9977_T_C |  8 |  4 | 1.750000 |  6 | 0 | 2 | 0.25000000 | 0.0013612387 | <!----> | ⋯ | 0.03 | NA | <!----> | NA      | MT-CO3 | Y257Y | Synonymous | NA                            | T_C | transition   |\n",
       "| 9978_T_C |  4 | 21 | 7.500000 |  1 | 1 | 2 | 0.75000000 | 0.0006806194 | <!----> | ⋯ |   NA | NA | <!----> | NA      | MT-CO3 | W258R | Missense   | NA                            | T_C | transition   |\n",
       "| 9985_G_A | 14 |  3 | 1.142857 | 13 | 0 | 1 | 0.07142857 | 0.0023821678 | <!----> | ⋯ | 0.00 | NA | <!----> | NA      | MT-CO3 | G260E | Missense   | NA                            | G_A | transition   |\n",
       "| 9986_G_A | 10 |  7 | 1.600000 |  9 | 0 | 1 | 0.10000000 | 0.0017015484 | <!----> | ⋯ | 0.21 |  1 | <!----> | NA      | MT-CO3 | G260G | Synonymous | NA                            | G_A | transition   |\n",
       "| 9987_T_C |  5 | 11 | 3.200000 |  3 | 1 | 1 | 0.40000000 | 0.0008507742 | <!----> | ⋯ | 0.00 | NA | <!----> | NA      | MT-CO3 | S261P | Missense   | NA                            | T_C | transition   |\n",
       "| 9993_T_C |  4 | 10 | 3.750000 |  2 | 0 | 2 | 0.50000000 | 0.0006806194 | <!----> | ⋯ |   NA | NA | <!----> | NA      | NA     | NA    | NA         | NA                            | T_C | transition   |\n",
       "| 9998_T_C |  2 |  3 | 2.000000 |  1 | 0 | 1 | 0.50000000 | 0.0003403097 | <!----> | ⋯ |   NA | NA | <!----> | T_run_4 | NA     | NA    | NA         | NA                            | T_C | transition   |\n",
       "\n"
      ],
      "text/plain": [
       "     Variants  CellN max_count mean_count ct1_cellN ct2_cellN ct3plus_cellN\n",
       "1    1000_T_C   3     4         2.000000  2         0         1            \n",
       "2    10005_A_G  2     3         2.000000  1         0         1            \n",
       "3    10007_T_C  2     3         2.500000  0         1         1            \n",
       "4    10009_G_A  2    10         5.500000  1         0         1            \n",
       "5    10010_T_C  3    12         4.666667  2         0         1            \n",
       "6    10020_T_C  6    13         4.833333  1         2         3            \n",
       "7    10031_T_C  3     2         1.333333  2         1         0            \n",
       "8    1004_G_A   4     3         2.000000  2         0         2            \n",
       "9    10045_T_C  2    32        16.500000  1         0         1            \n",
       "10   1005_T_C   6     7         2.000000  5         0         1            \n",
       "11   10054_G_A  7    11         2.428571  6         0         1            \n",
       "12   1006_T_C   9    30         4.555556  7         0         2            \n",
       "13   10063_A_G  4     4         2.750000  1         1         2            \n",
       "14   10066_T_C  4     5         2.000000  3         0         1            \n",
       "15   10068_G_A 10     4         1.900000  6         1         3            \n",
       "16   10071_T_C  4     2         1.250000  3         1         0            \n",
       "17   10075_T_C  2    12         6.500000  1         0         1            \n",
       "18   10078_T_C  2     5         3.000000  1         0         1            \n",
       "19   1008_A_G   2     2         1.500000  1         1         0            \n",
       "20   10081_T_C  3    33        11.666667  2         0         1            \n",
       "21   10083_A_G  2     5         3.000000  1         0         1            \n",
       "22   10084_T_C  4     8         3.250000  2         0         2            \n",
       "23   10088_C_T  6     7         2.000000  5         0         1            \n",
       "24   10093_T_C  9     5         1.444444  8         0         1            \n",
       "25   10098_G_A  7    13         3.000000  5         0         2            \n",
       "26   10101_T_C 10     4         1.500000  8         0         2            \n",
       "27   10102_T_C  2     3         2.000000  1         0         1            \n",
       "28   10114_T_C  7    28        12.285714  2         0         5            \n",
       "29   10115_T_C  4    18         6.250000  1         1         2            \n",
       "30   10126_G_A  5     7         2.200000  4         0         1            \n",
       "⋮    ⋮         ⋮     ⋮         ⋮          ⋮         ⋮         ⋮            \n",
       "3753 9914_A_G  13    36        3.692308   12        0         1            \n",
       "3754 9915_G_A  12    11        2.000000    9        2         1            \n",
       "3755 9918_G_A  51    12        1.235294   49        1         1            \n",
       "3756 9921_G_A   7    11        2.428571    6        0         1            \n",
       "3757 9925_G_A   8     5        1.875000    6        0         2            \n",
       "3758 9927_T_C   4    19        6.750000    1        1         2            \n",
       "3759 9928_A_G   6     4        1.500000    5        0         1            \n",
       "3760 9932_G_A  15    32        3.533333   12        0         3            \n",
       "3761 9934_A_G   2     3        2.000000    1        0         1            \n",
       "3762 9938_T_C   4     5        2.250000    2        1         1            \n",
       "3763 9939_G_A  10    10        2.800000    7        0         3            \n",
       "3764 9940_T_C   2    12        7.000000    0        1         1            \n",
       "3765 9942_G_A  15    36        6.800000    8        0         7            \n",
       "3766 9947_G_A  10    32        4.500000    6        2         2            \n",
       "3767 9948_G_A  10    17        5.000000    6        0         4            \n",
       "3768 9948_G_C   2    10        5.500000    1        0         1            \n",
       "3769 9949_T_C   2     7        4.000000    1        0         1            \n",
       "3770 9950_T_C  13     6        1.384615   12        0         1            \n",
       "3771 9964_A_G   6    13        5.166667    3        0         3            \n",
       "3772 9965_T_C   3     4        2.000000    2        0         1            \n",
       "3773 9966_G_A   6     8        2.833333    4        0         2            \n",
       "3774 9967_T_C   4     3        1.500000    3        0         1            \n",
       "3775 9973_T_C   7    18        4.000000    5        0         2            \n",
       "3776 9977_T_C   8     4        1.750000    6        0         2            \n",
       "3777 9978_T_C   4    21        7.500000    1        1         2            \n",
       "3778 9985_G_A  14     3        1.142857   13        0         1            \n",
       "3779 9986_G_A  10     7        1.600000    9        0         1            \n",
       "3780 9987_T_C   5    11        3.200000    3        1         1            \n",
       "3781 9993_T_C   4    10        3.750000    2        0         2            \n",
       "3782 9998_T_C   2     3        2.000000    1        0         1            \n",
       "     pct_ct2plus cellNPCT     hyper_label ⋯ freq_eurasian n_haplos\n",
       "1    0.3333333   0.0005104645             ⋯   NA          NA      \n",
       "2    0.5000000   0.0003403097             ⋯ 0.02          NA      \n",
       "3    1.0000000   0.0003403097             ⋯ 0.10          NA      \n",
       "4    0.5000000   0.0003403097             ⋯   NA          NA      \n",
       "5    0.3333333   0.0005104645             ⋯   NA          NA      \n",
       "6    0.8333333   0.0010209290             ⋯ 0.01          NA      \n",
       "7    0.3333333   0.0005104645             ⋯ 0.18           2      \n",
       "8    0.5000000   0.0006806194             ⋯   NA          NA      \n",
       "9    0.5000000   0.0003403097             ⋯ 0.04          NA      \n",
       "10   0.1666667   0.0010209290             ⋯ 0.62           8      \n",
       "11   0.1428571   0.0011910839             ⋯   NA          NA      \n",
       "12   0.2222222   0.0015313936             ⋯ 0.00          NA      \n",
       "13   0.7500000   0.0006806194             ⋯   NA          NA      \n",
       "14   0.2500000   0.0006806194             ⋯   NA          NA      \n",
       "15   0.4000000   0.0017015484             ⋯   NA          NA      \n",
       "16   0.2500000   0.0006806194             ⋯ 0.00          NA      \n",
       "17   0.5000000   0.0003403097             ⋯   NA          NA      \n",
       "18   0.5000000   0.0003403097             ⋯   NA          NA      \n",
       "19   0.5000000   0.0003403097             ⋯ 0.07          NA      \n",
       "20   0.3333333   0.0005104645             ⋯ 0.00          NA      \n",
       "21   0.5000000   0.0003403097             ⋯ 0.00          NA      \n",
       "22   0.5000000   0.0006806194             ⋯ 0.83           1      \n",
       "23   0.1666667   0.0010209290             ⋯ 0.04          NA      \n",
       "24   0.1111111   0.0015313936             ⋯   NA          NA      \n",
       "25   0.2857143   0.0011910839             ⋯   NA          NA      \n",
       "26   0.2000000   0.0017015484             ⋯ 0.25           1      \n",
       "27   0.5000000   0.0003403097             ⋯   NA          NA      \n",
       "28   0.7142857   0.0011910839             ⋯ 0.01          NA      \n",
       "29   0.7500000   0.0006806194             ⋯ 0.03           5      \n",
       "30   0.2000000   0.0008507742             ⋯   NA          NA      \n",
       "⋮    ⋮           ⋮            ⋮           ⋱ ⋮             ⋮       \n",
       "3753 0.07692308  0.0022120129             ⋯ 0.00          NA      \n",
       "3754 0.25000000  0.0020418581             ⋯   NA          NA      \n",
       "3755 0.03921569  0.0086778969             ⋯   NA          NA      \n",
       "3756 0.14285714  0.0011910839             ⋯ 0.13           1      \n",
       "3757 0.25000000  0.0013612387             ⋯   NA          NA      \n",
       "3758 0.75000000  0.0006806194             ⋯   NA          NA      \n",
       "3759 0.16666667  0.0010209290             ⋯   NA          NA      \n",
       "3760 0.20000000  0.0025523226             ⋯ 0.40           2      \n",
       "3761 0.50000000  0.0003403097             ⋯   NA          NA      \n",
       "3762 0.50000000  0.0006806194             ⋯ 0.13          NA      \n",
       "3763 0.30000000  0.0017015484             ⋯   NA          NA      \n",
       "3764 1.00000000  0.0003403097             ⋯   NA          NA      \n",
       "3765 0.46666667  0.0025523226             ⋯   NA          NA      \n",
       "3766 0.40000000  0.0017015484             ⋯ 1.13           2      \n",
       "3767 0.40000000  0.0017015484             ⋯ 0.21          NA      \n",
       "3768 0.50000000  0.0003403097             ⋯   NA          NA      \n",
       "3769 0.50000000  0.0003403097             ⋯   NA          NA      \n",
       "3770 0.07692308  0.0022120129             ⋯ 4.73          17      \n",
       "3771 0.50000000  0.0010209290             ⋯   NA          NA      \n",
       "3772 0.33333333  0.0005104645             ⋯ 0.01          NA      \n",
       "3773 0.33333333  0.0010209290             ⋯ 0.57           5      \n",
       "3774 0.25000000  0.0006806194             ⋯ 0.01          NA      \n",
       "3775 0.28571429  0.0011910839             ⋯   NA          NA      \n",
       "3776 0.25000000  0.0013612387             ⋯ 0.03          NA      \n",
       "3777 0.75000000  0.0006806194             ⋯   NA          NA      \n",
       "3778 0.07142857  0.0023821678             ⋯ 0.00          NA      \n",
       "3779 0.10000000  0.0017015484             ⋯ 0.21           1      \n",
       "3780 0.40000000  0.0008507742             ⋯ 0.00          NA      \n",
       "3781 0.50000000  0.0006806194             ⋯   NA          NA      \n",
       "3782 0.50000000  0.0003403097             ⋯   NA          NA      \n",
       "     blacklist_region homopolymer gene   aachange impact    \n",
       "1                     NA          NA     NA       NA        \n",
       "2                     NA          NA     NA       NA        \n",
       "3                     NA          NA     NA       NA        \n",
       "4                     NA          NA     NA       NA        \n",
       "5                     NA          NA     NA       NA        \n",
       "6                     NA          NA     NA       NA        \n",
       "7                     NA          NA     NA       NA        \n",
       "8                     NA          NA     NA       NA        \n",
       "9                     NA          NA     NA       NA        \n",
       "10                    NA          NA     NA       NA        \n",
       "11                    NA          NA     NA       NA        \n",
       "12                    NA          NA     NA       NA        \n",
       "13                    NA          MT-ND3 N2S      Missense  \n",
       "14                    NA          MT-ND3 F3S      Missense  \n",
       "15                    NA          MT-ND3 A4T      Missense  \n",
       "16                    NA          MT-ND3 L5L      Synonymous\n",
       "17                    T_run_4     MT-ND3 I6T      Missense  \n",
       "18                    T_run_4     MT-ND3 L7S      Missense  \n",
       "19                    NA          NA     NA       NA        \n",
       "20                    NA          MT-ND3 M8T      Missense  \n",
       "21                    NA          MT-ND3 I9V      Missense  \n",
       "22                    NA          MT-ND3 I9T      Missense  \n",
       "23                    NA          MT-ND3 N10N     Synonymous\n",
       "24                    NA          MT-ND3 L12P     Missense  \n",
       "25                    NA          MT-ND3 A14T     Missense  \n",
       "26                    NA          MT-ND3 L15L     Synonymous\n",
       "27                    NA          MT-ND3 L15S     Missense  \n",
       "28                    NA          MT-ND3 I19T     Missense  \n",
       "29                    NA          MT-ND3 I19I     Synonymous\n",
       "30                    NA          MT-ND3 W23*     Nonsense  \n",
       "⋮    ⋮                ⋮           ⋮      ⋮        ⋮         \n",
       "3753                  NA          MT-CO3 E236E    Synonymous\n",
       "3754                  NA          MT-CO3 A237T    Missense  \n",
       "3755                  NA          MT-CO3 A238T    Missense  \n",
       "3756                  NA          MT-CO3 A239T    Missense  \n",
       "3757                  NA          MT-CO3 W240*    Nonsense  \n",
       "3758                  NA          MT-CO3 Y241H    Missense  \n",
       "3759                  NA          MT-CO3 Y241C    Missense  \n",
       "3760                  NA          MT-CO3 W242W    Synonymous\n",
       "3761                  T_run_4     MT-CO3 H243R    Missense  \n",
       "3762                  T_run_4     MT-CO3 F244F    Synonymous\n",
       "3763                  NA          MT-CO3 V245M    Missense  \n",
       "3764                  NA          MT-CO3 V245A    Missense  \n",
       "3765                  NA          MT-CO3 D246N    Missense  \n",
       "3766                  NA          MT-CO3 V247V    Synonymous\n",
       "3767                  NA          MT-CO3 V248I    Missense  \n",
       "3768                  NA          MT-CO3 V248L    Missense  \n",
       "3769                  NA          MT-CO3 V248A    Missense  \n",
       "3770                  NA          MT-CO3 V248V    Synonymous\n",
       "3771                  NA          MT-CO3 Y253C    Missense  \n",
       "3772                  NA          MT-CO3 Y253Y    Synonymous\n",
       "3773                  NA          MT-CO3 V254I    Missense  \n",
       "3774                  NA          MT-CO3 V254A    Missense  \n",
       "3775                  NA          MT-CO3 I256T    Missense  \n",
       "3776                  NA          MT-CO3 Y257Y    Synonymous\n",
       "3777                  NA          MT-CO3 W258R    Missense  \n",
       "3778                  NA          MT-CO3 G260E    Missense  \n",
       "3779                  NA          MT-CO3 G260G    Synonymous\n",
       "3780                  NA          MT-CO3 S261P    Missense  \n",
       "3781                  NA          NA     NA       NA        \n",
       "3782                  T_run_4     NA     NA       NA        \n",
       "     Disease                       changes types       \n",
       "1    NA                            T_C     transition  \n",
       "2    NA                            A_G     transition  \n",
       "3    NA                            T_C     transition  \n",
       "4    NA                            G_A     transition  \n",
       "5    NA                            T_C     transition  \n",
       "6    NA                            T_C     transition  \n",
       "7    NA                            T_C     transition  \n",
       "8    NA                            G_A     transition  \n",
       "9    NA                            T_C     transition  \n",
       "10   NA                            T_C     transition  \n",
       "11   NA                            G_A     transition  \n",
       "12   NA                            T_C     transition  \n",
       "13   NA                            A_G     transition  \n",
       "14   NA                            T_C     transition  \n",
       "15   NA                            G_A     transition  \n",
       "16   NA                            T_C     transition  \n",
       "17   NA                            T_C     transition  \n",
       "18   NA                            T_C     transition  \n",
       "19   NA                            A_G     transition  \n",
       "20   NA                            T_C     transition  \n",
       "21   NA                            A_G     transition  \n",
       "22   NA                            T_C     transition  \n",
       "23   NA                            C_T     transition  \n",
       "24   NA                            T_C     transition  \n",
       "25   NA                            G_A     transition  \n",
       "26   NA                            T_C     transition  \n",
       "27   NA                            T_C     transition  \n",
       "28   NA                            T_C     transition  \n",
       "29   NA                            T_C     transition  \n",
       "30   NA                            G_A     transition  \n",
       "⋮    ⋮                             ⋮       ⋮           \n",
       "3753 NA                            A_G     transition  \n",
       "3754 NA                            G_A     transition  \n",
       "3755 NA                            G_A     transition  \n",
       "3756 Possible LHON helper mutation G_A     transition  \n",
       "3757 NA                            G_A     transition  \n",
       "3758 NA                            T_C     transition  \n",
       "3759 NA                            A_G     transition  \n",
       "3760 NA                            G_A     transition  \n",
       "3761 NA                            A_G     transition  \n",
       "3762 NA                            T_C     transition  \n",
       "3763 NA                            G_A     transition  \n",
       "3764 NA                            T_C     transition  \n",
       "3765 NA                            G_A     transition  \n",
       "3766 NA                            G_A     transition  \n",
       "3767 NA                            G_A     transition  \n",
       "3768 NA                            G_C     transversion\n",
       "3769 NA                            T_C     transition  \n",
       "3770 NA                            T_C     transition  \n",
       "3771 NA                            A_G     transition  \n",
       "3772 NA                            T_C     transition  \n",
       "3773 LHON possible helper variant  G_A     transition  \n",
       "3774 NA                            T_C     transition  \n",
       "3775 NA                            T_C     transition  \n",
       "3776 NA                            T_C     transition  \n",
       "3777 NA                            T_C     transition  \n",
       "3778 NA                            G_A     transition  \n",
       "3779 NA                            G_A     transition  \n",
       "3780 NA                            T_C     transition  \n",
       "3781 NA                            T_C     transition  \n",
       "3782 NA                            T_C     transition  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "annotate_all_variants(variant_summary_stat_from_mtx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e58991ab-d22f-441e-9a68-7a2ec46519d6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.3",
   "language": "R",
   "name": "ir43"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
