<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ redeemR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">redeemR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ATAC_Wrapper.html">ATAC_Wrapper()</a></code> </p>
        </td>
        <td><p>Wrap Seurat ATAC clustering</p></td>
      </tr><tr><td>
          <p><code><a href="AddDatatoplot_clustering-redeemR-method.html">AddDatatoplot_clustering(<i>&lt;redeemR&gt;</i>)</a></code> </p>
        </td>
        <td><p>AddDatatoplot_clustering
This prepare the clonal clustering data to plot</p></td>
      </tr><tr><td>
          <p><code><a href="AddDatatoplot_clustering.html">AddDatatoplot_clustering()</a></code> </p>
        </td>
        <td><p>AddDatatoplot_clustering
This prepare the clonal clustering data to plot</p></td>
      </tr><tr><td>
          <p><code><a href="AddDist-redeemR-method.html">AddDist(<i>&lt;redeemR&gt;</i>)</a></code> </p>
        </td>
        <td><p>AddDist
This add Jaccard, Dice, Jaccard3W distance and stored in DistObjects</p></td>
      </tr><tr><td>
          <p><code><a href="AddDist.html">AddDist()</a></code> </p>
        </td>
        <td><p>AddDist
This add Jaccard, Dice, Jaccard3W distance and stored in DistObjects</p></td>
      </tr><tr><td>
          <p><code><a href="AddHemSignature.html">AddHemSignature()</a></code> </p>
        </td>
        <td><p>Function to add hematopoietic signatures from Griffin_Signatures</p></td>
      </tr><tr><td>
          <p><code><a href="AddTree-redeemR-method.html">AddTree(<i>&lt;redeemR&gt;</i>)</a></code> </p>
        </td>
        <td><p>Add_Tree
Optional, if a phylogentic tree object phylo is already available, can be directly added to the redeemR class in slot TREE</p></td>
      </tr><tr><td>
          <p><code><a href="AddTree.html">AddTree()</a></code> </p>
        </td>
        <td><p>Add_Tree
Optional, if a phylogentic tree object phylo is already available, can be directly added to the redeemR</p></td>
      </tr><tr><td>
          <p><code><a href="Add_AssignVariant-redeemR-method.html">Add_AssignVariant(<i>&lt;redeemR&gt;</i>)</a></code> </p>
        </td>
        <td><p>a function to assign variants to edges based on maximum likihood</p></td>
      </tr><tr><td>
          <p><code><a href="Add_AssignVariant.html">Add_AssignVariant()</a></code> </p>
        </td>
        <td><p>Add_AssignVariant
a function to assign variants to edges based on maximum likihood</p></td>
      </tr><tr><td>
          <p><code><a href="Add_DepthMatrix-redeemR-method.html">Add_DepthMatrix(<i>&lt;redeemR&gt;</i>)</a></code> </p>
        </td>
        <td><p>Add_DepthMatrix
Optional, add a matrix with same dimension with the Cts.Mtx and Cts.Mtx.bi, which display the depths</p></td>
      </tr><tr><td>
          <p><code><a href="Add_DepthMatrix.html">Add_DepthMatrix()</a></code> </p>
        </td>
        <td><p>Add_DepthMatrix
Optional, add a matrix with same dimension with the Cts.Mtx and Cts.Mtx.bi, which display the depths</p></td>
      </tr><tr><td>
          <p><code><a href="Add_tree_cut-redeemR-method.html">Add_tree_cut(<i>&lt;redeemR&gt;</i>)</a></code> </p>
        </td>
        <td><p>a function to cut tree using assigned variant as branch-length on edge</p></td>
      </tr><tr><td>
          <p><code><a href="Add_tree_cut.html">Add_tree_cut()</a></code> </p>
        </td>
        <td><p>Add_tree_cut
a function to cut tree using assigned variant as branch-length on edge</p></td>
      </tr><tr><td>
          <p><code><a href="BinaryDist.html">BinaryDist()</a></code> </p>
        </td>
        <td><p>Compute distances for binary distances</p></td>
      </tr><tr><td>
          <p><code><a href="CW_mgatk.read.html">CW_mgatk.read()</a></code> </p>
        </td>
        <td><p>Function to read in mitoV outputs</p></td>
      </tr><tr><td>
          <p><code><a href="Clone_FinderMarker.html">Clone_FinderMarker()</a></code> </p>
        </td>
        <td><p>Define a function to perform Find marker for top vs bottom clones
This function was developed based on DN4T2.basics.ipynb</p></td>
      </tr><tr><td>
          <p><code><a href="ComputeRejectRate.html">ComputeRejectRate()</a></code> </p>
        </td>
        <td><p>Function to compute the reject rate(The filtering rate in concensus variant calling)</p></td>
      </tr><tr><td>
          <p><code><a href="CountVperCell.html">CountVperCell()</a></code> </p>
        </td>
        <td><p>Internal function in plot_variant</p></td>
      </tr><tr><td>
          <p><code><a href="Create_redeemR.html">Create_redeemR()</a></code> </p>
        </td>
        <td><p>Create_redeemR</p></td>
      </tr><tr><td>
          <p><code><a href="DE.gettripple.html">DE.gettripple()</a></code> </p>
        </td>
        <td><p>DE.gettripple</p></td>
      </tr><tr><td>
          <p><code><a href="Datatoplots-class.html">Datatoplots-class</a></code> <code><a href="Datatoplots-class.html">Datatoplots</a></code> </p>
        </td>
        <td><p>An intermediate  S4 class Datatoplots</p></td>
      </tr><tr><td>
          <p><code><a href="DepthSummary.html">DepthSummary()</a></code> </p>
        </td>
        <td><p>Function to summarize the depth (Total that passed Q30)</p></td>
      </tr><tr><td>
          <p><code><a href="DistObjects-class.html">DistObjects-class</a></code> <code><a href="DistObjects-class.html">DistObjects</a></code> </p>
        </td>
        <td><p>An intermediate  S4 class Datatoplots</p></td>
      </tr><tr><td>
          <p><code><a href="DoDE.html">DoDE()</a></code> </p>
        </td>
        <td><p>DoDE</p></td>
      </tr><tr><td>
          <p><code><a href="FromDist2Graph.html">FromDist2Graph()</a></code> </p>
        </td>
        <td><p>FromDist2Graph
From disttance object or matrix to graph, default is to return igraph object
This function was developed based on</p></td>
      </tr><tr><td>
          <p><code><a href="GEM_Wrapper.html">GEM_Wrapper()</a></code> </p>
        </td>
        <td><p>Wrap Seurat RNA clustering</p></td>
      </tr><tr><td>
          <p><code><a href="GTSummary.html">GTSummary()</a></code> </p>
        </td>
        <td><p>Function to generate GTS summary</p></td>
      </tr><tr><td>
          <p><code><a href="LineageBiasPlot.html">LineageBiasPlot()</a></code> </p>
        </td>
        <td><p>plot_npSummary to plot the lineage composition</p></td>
      </tr><tr><td>
          <p><code><a href="MakeAllNodes.html">MakeAllNodes()</a></code> </p>
        </td>
        <td><p>Define a function make the Allnodes(Node|Parent|Freq|CladeSize), where Freq is the number of variants assigned to the node(as ending point)
from redeemR object,</p></td>
      </tr><tr><td>
          <p><code><a href="MakeDF4Regress.html">MakeDF4Regress()</a></code> </p>
        </td>
        <td><p>MakeDF4Regress
Define a function to make two dataframe for regression analysis
This function was developed based on HSC_multiome_Het_2.ipynb</p></td>
      </tr><tr><td>
          <p><code><a href="MakeNN.html">MakeNN()</a></code> </p>
        </td>
        <td><p>Define a function to make nn list, which can be further used to make adjacency matrix
This scan row by row, looking for k.param nearest neighbours</p></td>
      </tr><tr><td>
          <p><code><a href="Make_AnnTable.html">Make_AnnTable()</a></code> </p>
        </td>
        <td><p>Make_AnnTable, Make a big dataframe, each row is a cell, each column includes info such as clonal UMAP, Clonal ID, ATAC/RNA/WNN UMAP, PCA, gene expression of chosen gene, etc.  Require a redeemR object and a multiome wrapper that better matches the cells in the redeemR</p></td>
      </tr><tr><td>
          <p><code><a href="Make_Cells4Nodes.html">Make_Cells4Nodes()</a></code> </p>
        </td>
        <td><p>Define a function to make a list, each contains the cell names for a node</p></td>
      </tr><tr><td>
          <p><code><a href="Make_matrix-redeemR-method.html">Make_matrix(<i>&lt;redeemR&gt;</i>)</a></code> </p>
        </td>
        <td><p>Make_matrix
This will make the matixies of Cell VS mitochondrial variants and return redeemR
Results stored in Cts.Mtx and Cts.Mtx.bi</p></td>
      </tr><tr><td>
          <p><code><a href="Make_matrix.html">Make_matrix()</a></code> </p>
        </td>
        <td><p>Make_matrix
This will make the matixies of Cell VS mitochondrial variants and return redeemR
Results stored in Cts.Mtx and Cts.Mtx.bi</p></td>
      </tr><tr><td>
          <p><code><a href="Make_tree-redeemR-method.html">Make_tree(<i>&lt;redeemR&gt;</i>)</a></code> </p>
        </td>
        <td><p>Make_tree
This will generate a basic phylogenetic tree</p></td>
      </tr><tr><td>
          <p><code><a href="Make_tree.html">Make_tree()</a></code> </p>
        </td>
        <td><p>Make_tree
This will generate a basic phylogenetic tree</p></td>
      </tr><tr><td>
          <p><code><a href="MergeMtx.html">MergeMtx()</a></code> </p>
        </td>
        <td><p>Function to Merge sparse Matrix</p></td>
      </tr><tr><td>
          <p><code><a href="Motifenrich.binom.html">Motifenrich.binom()</a></code> </p>
        </td>
        <td><p>Motifenrich.binom
In house function to compute enrichment from Fimo
This function was developed based on HSC_multiome_Het.ipynb and HSC_multiome_Het_2.ipynb</p></td>
      </tr><tr><td>
          <p><code><a href="Multi_Wrapper.html">Multi_Wrapper()</a></code> </p>
        </td>
        <td><p>Wrap Seurat Multiomics clustering</p></td>
      </tr><tr><td>
          <p><code><a href="MutationProfile.bulk.html">MutationProfile.bulk()</a></code> </p>
        </td>
        <td><p>Function to plot bulk level mutation signatures</p></td>
      </tr><tr><td>
          <p><code><a href="NN2M.html">NN2M()</a></code> </p>
        </td>
        <td><p>Define a function convert nn list to adjacency matrix that can be further used for igraph</p></td>
      </tr><tr><td>
          <p><code><a href="ProgenyMapping.html">ProgenyMapping()</a></code> </p>
        </td>
        <td><p>Define a function to perform single-cell based hard porogeny assignment
This function was developed based on DN4T2.basics.ipynb</p></td>
      </tr><tr><td>
          <p><code><a href="ProgenyMapping_np.html">ProgenyMapping_np()</a></code> </p>
        </td>
        <td><p>ProgenyMapping_np
Define a function to compute network propagation based probability
FromDist2Graph is needed to convert fistance matrix into MNN graph</p></td>
      </tr><tr><td>
          <p><code><a href="Reclustering.html">Reclustering()</a></code> </p>
        </td>
        <td><p>Function to reclustering a seurat object</p></td>
      </tr><tr><td>
          <p><code><a href="Reclustering_hm.html">Reclustering_hm()</a></code> </p>
        </td>
        <td><p>Function to reclustering_hm a seurat object with Harmony</p></td>
      </tr><tr><td>
          <p><code><a href="Run_Lin_regression.html">Run_Lin_regression()</a></code> </p>
        </td>
        <td><p>Run_Lin_regression</p></td>
      </tr><tr><td>
          <p><code><a href="Run_Lin_regression_poi.html">Run_Lin_regression_poi()</a></code> </p>
        </td>
        <td><p>Run_Lin_regression_poi
Firstly used in HSC_multiome_Het_2.ipynb
This function was developed based on</p></td>
      </tr><tr><td>
          <p><code><a href="Runplot_scale_2.html">Runplot_scale_2()</a></code> </p>
        </td>
        <td><p>plot_npSummary to assess the outputlevel vs lineage bias, normalize by assigned</p></td>
      </tr><tr><td>
          <p><code><a href="Runplot_scale_3.html">Runplot_scale_3()</a></code> </p>
        </td>
        <td><p>plot_npSummary to assess the outputlevel vs lineage bias, normalize by HSC original clone size</p></td>
      </tr><tr><td>
          <p><code><a href="SeuratLSIClustering-redeemR-method.html">SeuratLSIClustering(<i>&lt;redeemR&gt;</i>)</a></code> </p>
        </td>
        <td><p>SeuratLSIClustering
This will use the mito variants for Seurat clustering (LSI based)</p></td>
      </tr><tr><td>
          <p><code><a href="SeuratLSIClustering.html">SeuratLSIClustering()</a></code> </p>
        </td>
        <td><p>SeuratLSIClustering
This will use the mito variants for Seurat clustering (LSI based)</p></td>
      </tr><tr><td>
          <p><code><a href="Subset_redeemR.html">Subset_redeemR()</a></code> </p>
        </td>
        <td><p>Subset_redeemR Subset a redeemR object by selecting a subset of cells, return a new redeemR object with only 4 slots: para; CellMeta; Cts.Mtx.bi; UniqueV, can be used for downstreme compute distance, clonal clustering, make tree, etc</p></td>
      </tr><tr><td>
          <p><code><a href="TREE-class.html">TREE-class</a></code> <code><a href="TREE-class.html">TREE</a></code> </p>
        </td>
        <td><p>An intermediate  S4 class Tree that store tree info</p></td>
      </tr><tr><td>
          <p><code><a href="Tomerge_v2.html">Tomerge_v2()</a></code> </p>
        </td>
        <td><p>Tomerge_v2</p></td>
      </tr><tr><td>
          <p><code><a href="Translate_RNA2ATAC.html">Translate_RNA2ATAC()</a></code> </p>
        </td>
        <td><p>Function to translate the RNA barcode into ATAC barcode and add a column</p></td>
      </tr><tr><td>
          <p><code><a href="Translate_simple_ATAC2RNA.html">Translate_simple_ATAC2RNA()</a></code> </p>
        </td>
        <td><p>Translate_simple_ATAC2RNA</p></td>
      </tr><tr><td>
          <p><code><a href="Translate_simple_RNA2ATAC.html">Translate_simple_RNA2ATAC()</a></code> </p>
        </td>
        <td><p>Translate_simple_RNA2ATAC</p></td>
      </tr><tr><td>
          <p><code><a href="Vfilter_v3.html">Vfilter_v3()</a></code> </p>
        </td>
        <td><p>Function to filter variants</p></td>
      </tr><tr><td>
          <p><code><a href="add_derived_profile_info.html">add_derived_profile_info()</a></code> </p>
        </td>
        <td><p>This is a convinience function, internal borrowed from https://github.com/NickWilliamsSanger/treemut/blob/main/R/treemut.R#L68</p></td>
      </tr><tr><td>
          <p><code><a href="df2ProfileMtx.html">df2ProfileMtx()</a></code> </p>
        </td>
        <td><p>This is a convinience function, internal</p></td>
      </tr><tr><td>
          <p><code><a href="get_ancestral_nodes.html">get_ancestral_nodes()</a></code> </p>
        </td>
        <td><p>This is a convinience function, internal borrowed from https://github.com/NickWilliamsSanger/treemut/blob/main/R/treemut.R#L68</p></td>
      </tr><tr><td>
          <p><code><a href="plot_depth.html">plot_depth()</a></code> </p>
        </td>
        <td><p>Function to plot the mito depth summary</p></td>
      </tr><tr><td>
          <p><code><a href="plot_npSummary.html">plot_npSummary()</a></code> </p>
        </td>
        <td><p>plot_npSummary to assess the outputlevel</p></td>
      </tr><tr><td>
          <p><code><a href="plot_variant.html">plot_variant()</a></code> </p>
        </td>
        <td><p>Function to plot variant metrics</p></td>
      </tr><tr><td>
          <p><code><a href="quick_w_cosine.html">quick_w_cosine()</a></code> </p>
        </td>
        <td><p>Compute weighted cosine distance</p></td>
      </tr><tr><td>
          <p><code><a href="quick_w_jaccard.html">quick_w_jaccard()</a></code> </p>
        </td>
        <td><p>Compute weighted jaccard distance</p></td>
      </tr><tr><td>
          <p><code><a href="reconstruct_genotype_summary.html">reconstruct_genotype_summary()</a></code> </p>
        </td>
        <td><p>This is a function borrowed from https://github.com/NickWilliamsSanger/treemut/blob/main/R/treemut.R#L68
Input phylo object, return a "profile matrix"--Edge(or denoted as the ending node) vs cell. a 0, 1 character string that indicate what cells in a given node</p></td>
      </tr><tr><td>
          <p><code><a href="redeemR-class.html">redeemR-class</a></code> <code><a href="redeemR-class.html">redeemR</a></code> </p>
        </td>
        <td><p>Major redeem class that store clonal-resolved multi-omics</p></td>
      </tr><tr><td>
          <p><code><a href="show-redeemR-method.html">show(<i>&lt;redeemR&gt;</i>)</a></code> </p>
        </td>
        <td><p>show
This will show the basics of redeemR class</p></td>
      </tr><tr><td>
          <p><code><a href="split_profile.html">split_profile()</a></code> </p>
        </td>
        <td><p>This is a convinience function, internal  borrowed from https://github.com/NickWilliamsSanger/treemut/blob/main/R/treemut.R#L68</p></td>
      </tr><tr><td>
          <p><code><a href="str2vector.html">str2vector()</a></code> </p>
        </td>
        <td><p>This is a convinience function, internal</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Chen Weng.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

