<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Make_AnnTable, Make a big dataframe, each row is a cell, each column includes info such as clonal UMAP, Clonal ID, ATAC/RNA/WNN UMAP, PCA, gene expression of chosen gene, etc.  Require a MitoTracing object and a multiome wrapper that better matches the cells in the MitoTracing — Make_AnnTable • redeemR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Make_AnnTable, Make a big dataframe, each row is a cell, each column includes info such as clonal UMAP, Clonal ID, ATAC/RNA/WNN UMAP, PCA, gene expression of chosen gene, etc.  Require a MitoTracing object and a multiome wrapper that better matches the cells in the MitoTracing — Make_AnnTable"><meta property="og:description" content="Make_AnnTable, Make a big dataframe, each row is a cell, each column includes info such as clonal UMAP, Clonal ID, ATAC/RNA/WNN UMAP, PCA, gene expression of chosen gene, etc.  Require a MitoTracing object and a multiome wrapper that better matches the cells in the MitoTracing"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">redeemR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/package_intro.html">package_intro</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make_AnnTable, Make a big dataframe, each row is a cell, each column includes info such as clonal UMAP, Clonal ID, ATAC/RNA/WNN UMAP, PCA, gene expression of chosen gene, etc.  Require a MitoTracing object and a multiome wrapper that better matches the cells in the MitoTracing</h1>
    
    <div class="hidden name"><code>Make_AnnTable.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Make_AnnTable, Make a big dataframe, each row is a cell, each column includes info such as clonal UMAP, Clonal ID, ATAC/RNA/WNN UMAP, PCA, gene expression of chosen gene, etc.  Require a MitoTracing object and a multiome wrapper that better matches the cells in the MitoTracing</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">Make_AnnTable</span><span class="op">(</span></span>
<span>  Mitotracing <span class="op">=</span> <span class="va">DN4_HSC_mitoTracing.Sensitive</span>,</span>
<span>  Multiome <span class="op">=</span> <span class="va">Donor04_HSC_Multiome_wrapper</span>,</span>
<span>  clonal_features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCount_mitoV"</span>, <span class="st">"seurat_clusters"</span><span class="op">)</span>,</span>
<span>  clonal_features_rename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nCount_mitoV"</span>, <span class="st">"clone_clusters"</span><span class="op">)</span>,</span>
<span>  CellMeta_features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"meanCov"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_ATAC"</span>,</span>
<span>    <span class="st">"nFeature_ATAC"</span>, <span class="st">"CellType"</span><span class="op">)</span>,</span>
<span>  CellMeta_features_rename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Mito_meanCov"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"nFeature_RNA"</span>,</span>
<span>    <span class="st">"nCount_ATAC"</span>, <span class="st">"nFeature_ATAC"</span>, <span class="st">"CellType"</span><span class="op">)</span>,</span>
<span>  multiome_features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seurat_clusters"</span><span class="op">)</span>,</span>
<span>  multiome_features_rename <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NewSeurat_cluster"</span><span class="op">)</span>,</span>
<span>  RNAUMAP <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  ATACUMAP <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  WNNUMAP <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  PCA <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  LSI <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  Variants <span class="op">=</span> <span class="st">""</span>,</span>
<span>  genes <span class="op">=</span> <span class="st">""</span>,</span>
<span>  peaks <span class="op">=</span> <span class="st">""</span>,</span>
<span>  PostTrans_from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  PostTrans_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>Mitotracing</dt>
<dd><p>eg. DN4_HSC_mitoTracing.Sensitive</p></dd>


<dt>Multiome</dt>
<dd><p>eg. Donor04_HSC_Multiome_wrapper,  Multiome_wrapper object that matches with the MitoTracing, a reclustering using Multi_Wrapper() is recommended</p></dd>


<dt>clonal_features</dt>
<dd><p>eg. c("nCount_mitoV","seurat_clusters"), The column names take from Mitotracing@Seurat@meta.data, importantly the clonal clusterings</p></dd>


<dt>clonal_features_rename</dt>
<dd><p>eg. c("nCount_mitoV","clone_clusters") Rename the clonal_features</p></dd>


<dt>CellMeta_features</dt>
<dd><p>eg. c("meanCov","nCount_RNA","nFeature_RNA","nCount_ATAC","nFeature_ATAC","CellType") The column names take from  Mitotracing@CellMeta, may useful cell features</p></dd>


<dt>CellMeta_features_rename</dt>
<dd><p>eg. c("Mito_meanCov","nCount_RNA","nFeature_RNA","nCount_ATAC","nFeature_ATAC","CellType") Rename the CellMeta</p></dd>


<dt>multiome_features</dt>
<dd><p>eg. c("seurat_clusters")  The column names take from Multiome@meta.data</p></dd>


<dt>multiome_features_rename</dt>
<dd><p>eg. c("NewSeurat_cluster")   Rename the column names for multiome_features</p></dd>


<dt>RNAUMAP</dt>
<dd><p>default T</p></dd>


<dt>ATACUMAP</dt>
<dd><p>Default T</p></dd>


<dt>WNNUMAP</dt>
<dd><p>Default T</p></dd>


<dt>PCA</dt>
<dd><p>Default T</p></dd>


<dt>LSI</dt>
<dd><p>Default T</p></dd>


<dt>Variants</dt>
<dd><p>Default ""  can be a vector of variant names format is eg "Variants10020TC"</p></dd>


<dt>genes</dt>
<dd><p>Default ""  can be a vector of gene names, for example c("HLF","CD34")</p></dd>


<dt>peaks</dt>
<dd><p>Default ""  can be a vector of peaks names</p></dd>


<dt>PostTrans_from</dt>
<dd><p>Default c(2,3)  # This is a tricky part eh nmerging files are involved, find the postfix from cellranger agg for different sample</p></dd>


<dt>PostTrans_to</dt>
<dd><p>Default c(2,1)</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>AnnTable</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by First Last.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

